---
title: "DEG analysis"
output: html_document
date: "2023-10-01"
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/u/project/xyang123/julianas/medial_septum_sct")

library(limma)
library(edgeR)
library(data.table)
library(plyr)
library(dplyr)
library(fastDummies)
library(Seurat)
library(clusterProfiler)
library(enrichplot)
library(org.Mm.eg.db)
library(ggplot2)
library(RColorBrewer)
library(grid)
library(gridExtra)
library(SingleCellExperiment)
#library(Matrix.utils)
library(magrittr)
library(simplifyEnrichment)
library(stringr)
library(enrichplot)
library(EnhancedVolcano)
library(biomaRt)
library(readxl)
library(UpSetR)
library(tidyr)
library(purrr)
library(ggridges)
library(ggpubr)
library(ComplexHeatmap)
library(ComplexUpset)
library(ggplotify)
```

```{r}
seurat <- readRDS( "Saves/For_Figures.rds") %>% DietSeurat(counts = TRUE)
```

```{r}
gene_annotation <- read.table("Gene_annotation/gene_annotation_final2.txt", sep = "\t", header=TRUE)
par_genes <- c("Mid1", "Sts", "Nlgn4", "Asmt", "Mid1-ps", "Erdr1", "Mafl", "Asmtl", "Mafl-ps", "Dhrsx-ps", "VZI55075-ps", "Cd99", "Xg", "Arse", "Akap17a")
xescape_brain <- read_xlsx("Gene_annotation/xescape_brain.xlsx")
colnames(xescape_brain) <- xescape_brain[1, ]
xescape_brain <- xescape_brain[-1, ]

gene_annotation <- gene_annotation %>% mutate(PAR = ifelse(mgi_symbol %in% par_genes, "PAR", "Non-PAR"))
gene_annotation <- gene_annotation %>% mutate(Escapee = ifelse(mgi_symbol %in% xescape_brain$Gene, "Escapee", "Non-escapee"))
gene_annotation <- gene_annotation %>% mutate(Class = ifelse(chromosome_name %in% c("X", "Y"), "Sex-linked", "Autosomal"))
```

##Process DEG results into df 
```{r}
##old loop
allcells <- unique(seurat$cell.type)
allcells <- allcells[!allcells == "Unknown"]

alldata <- list.dirs("Data/Derived/DEG", recursive = TRUE)
alldata <- alldata[-c(1,2,3,4,5,7,8,13,14,15)]

topgenes <- list()
DEG_results <- list()

organism <- org.Mm.eg.db

gene_list <- rownames(seurat@assays$RNA@counts)
gene_list <- na.omit(gene_list)

for(i in 1:length(allcells)) {
  currentcell <- as.character(allcells[i])
  
  for(j in 1:length(alldata)) {
    load(paste0(alldata[j], "/", currentcell, ".rda"))
  }
  
  ##normative model
  for(k in 1:length(resultlist_normative)) {

    currentframe <- resultlist_normative[[k]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    #currentframe <- currentframe %>% dplyr::filter(adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcomparison <- names(resultlist_normative[k])
    currentmodel <- "Normative"
    currentframe$cell <- currentcell
    currentframe$model <- currentmodel
    currentframe$comparison <- currentcomparison
    currentframe <- currentframe %>% dplyr::arrange(-logFC)

    topgenes[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    DEG_results[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe
  }
  
  ##FCG model 
  for(k in 1:length(resultlist_fcg)) {
 
    currentframe <- resultlist_fcg[[k]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    #currentframe <- currentframe %>% dplyr::filter(adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcomparison <- names(resultlist_fcg[k])
    currentmodel <- "FCG"
    currentframe$cell <- currentcell
    currentframe$model <- currentmodel
    currentframe$comparison <- currentcomparison
    currentframe <- currentframe %>% dplyr::arrange(-logFC)

    topgenes[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    DEG_results[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe
  }
  
  ##FCG model dose
  for(k in 1:length(resultlist_fcg_dose)) {
 
    currentframe <- resultlist_fcg_dose[[k]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    #currentframe <- currentframe %>% dplyr::filter(adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcomparison <- names(resultlist_fcg_dose[k])
    currentmodel <- "FCG"
    currentframe$cell <- currentcell
    currentframe$model <- currentmodel
    currentframe$comparison <- currentcomparison
    currentframe <- currentframe %>% dplyr::arrange(-logFC)
    
    genes_to_filter <- resultlist_fcg$XX_vs_XY %>% tibble::rownames_to_column("gene") %>% dplyr::filter(adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!gene %in% genes_to_filter$gene)

    topgenes[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    DEG_results[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe
  }
  
  ##XY model (males)
  for(k in 1:length(resultlist_XY)) {
 
    currentframe <- resultlist_XY[[k]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    #currentframe <- currentframe %>% dplyr::filter(adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcomparison <- names(resultlist_XY[k])
    currentmodel <- "XY_males"
    currentframe$cell <- currentcell
    currentframe$model <- currentmodel
    currentframe$comparison <- currentcomparison
    currentframe <- currentframe %>% dplyr::arrange(-logFC)

    topgenes[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    DEG_results[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe
  }
  
   ##XY model (females)
  for(k in 1:length(resultlist_XY_female)) {
 
    currentframe <- resultlist_XY_female[[k]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    #currentframe <- currentframe %>% dplyr::filter(adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcomparison <- names(resultlist_XY_female[k])
    currentmodel <- "XY_females"
    currentframe$cell <- currentcell
    currentframe$model <- currentmodel
    currentframe$comparison <- currentcomparison
    currentframe <- currentframe %>% dplyr::arrange(-logFC)

    topgenes[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    DEG_results[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe
  }
  
  ##XY model (combined) 
  for(k in 1:2) {
    currentframe <- resultlist_XY_comb[[k]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    #currentframe <- currentframe %>% dplyr::filter(adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcomparison <- names(resultlist_XY_comb[k])
    currentmodel <- "XY_combined"
    currentframe$cell <- currentcell
    currentframe$model <- currentmodel
    currentframe$comparison <- currentcomparison
    currentframe <- currentframe %>% dplyr::arrange(-logFC)
    
    interaction_genes <- bind_rows(resultlist_XY_comb[3:6], .id = "comparison") %>% 
      dplyr::filter(adj.P.Val < 0.05) %>%
      tibble::rownames_to_column(var = "gene") %>%
      dplyr::select(gene) %>%
      unlist(use.names = FALSE)
    interaction_genes <- gsub("\\...*", "", interaction_genes)
    interaction_genes <- interaction_genes[!duplicated(interaction_genes)]
    
    currentframe <- currentframe %>% dplyr::filter(!gene %in% interaction_genes)

    topgenes[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    DEG_results[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe
  }
  
  ##full model 
  for(k in 1:length(resultlist_full)) {
 
    currentframe <- resultlist_full[[k]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    #currentframe <- currentframe %>% dplyr::filter(adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcomparison <- names(resultlist_full[k])
    currentmodel <- "Full"
    currentframe$cell <- currentcell
    currentframe$model <- currentmodel
    currentframe$comparison <- currentcomparison
    currentframe <- currentframe %>% dplyr::arrange(-logFC)

    topgenes[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    DEG_results[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe
  }
  
  ##full continuous model 
  for(k in 1:length(resultlist_full_cont)) {
 
    currentframe <- resultlist_full_cont[[k]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    #currentframe <- currentframe %>% dplyr::filter(adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcomparison <- names(resultlist_full_cont[k])
    currentmodel <- "Full_continuous"
    currentframe$cell <- currentcell
    currentframe$model <- currentmodel
    currentframe$comparison <- currentcomparison
    currentframe <- currentframe %>% dplyr::arrange(-logFC)

    topgenes[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    DEG_results[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe
  }
  
  ##AdditionalX model 
  for(k in 1:length(resultlist_additionalX)) {
 
    currentframe <- resultlist_additionalX[[k]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    #currentframe <- currentframe %>% dplyr::filter(adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcomparison <- names(resultlist_additionalX[k])
    currentmodel <- "additionalX"
    currentframe$cell <- currentcell
    currentframe$model <- currentmodel
    currentframe$comparison <- currentcomparison
    currentframe <- currentframe %>% dplyr::arrange(-logFC)

    topgenes[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    DEG_results[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe
  }
  
  ##AdditionalX continuous model 
  for(k in 1:length(resultlist_additionalX_cont)) {
 
    currentframe <- resultlist_additionalX_cont[[k]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    #currentframe <- currentframe %>% dplyr::filter(adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcomparison <- names(resultlist_additionalX_cont[k])
    currentmodel <- "additionalX_continuous"
    currentframe$cell <- currentcell
    currentframe$model <- currentmodel
    currentframe$comparison <- currentcomparison
    currentframe <- currentframe %>% dplyr::arrange(-logFC)

    topgenes[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    DEG_results[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe
  }
  
  ##AdditionalY model 
  for(k in 1:length(resultlist_additionalY)) {
 
    currentframe <- resultlist_additionalY[[k]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    #currentframe <- currentframe %>% dplyr::filter(adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcomparison <- names(resultlist_additionalY[k])
    currentmodel <- "additionalY"
    currentframe$cell <- currentcell
    currentframe$model <- currentmodel
    currentframe$comparison <- currentcomparison
    currentframe <- currentframe %>% dplyr::arrange(-logFC)

    topgenes[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    DEG_results[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe
  }
  
    ##AdditionalY model 
  for(k in 1:length(resultlist_additionalY_cont)) {
 
    currentframe <- resultlist_additionalY_cont[[k]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    #currentframe <- currentframe %>% dplyr::filter(adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcomparison <- names(resultlist_additionalY_cont[k])
    currentmodel <- "additionalY_continuous"
    currentframe$cell <- currentcell
    currentframe$model <- currentmodel
    currentframe$comparison <- currentcomparison
    currentframe <- currentframe %>% dplyr::arrange(-logFC)

    topgenes[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    DEG_results[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe
  }
}

save(topgenes, DEG_results, file = "Saves/DEG_results.Rda")
```

```{r}
allcells <- unique(seurat$cell.type)
allcells <- allcells[!allcells == "Unknown"]

alldata <- list.dirs("Data/Derived/DEG", recursive = TRUE)
alldata <- alldata[c(6,8,9,10,11,12)]

topgenes <- list()
DEG_results <- list()

organism <- org.Mm.eg.db

gene_list <- rownames(seurat@assays$RNA@counts)
gene_list <- na.omit(gene_list)

for(i in 1:length(allcells)) {
  currentcell <- as.character(allcells[i])
  
  for(j in 1:length(alldata)) {
    load(paste0(alldata[j], "/", currentcell, ".rda"))
  }
  
  ##normative model
  for(k in 1:length(resultlist_normative)) {

    currentframe <- resultlist_normative[[k]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    #currentframe <- currentframe %>% dplyr::filter(adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcomparison <- names(resultlist_normative[k])
    currentmodel <- "Normative"
    currentframe$cell <- currentcell
    currentframe$model <- currentmodel
    currentframe$comparison <- currentcomparison
    currentframe <- currentframe %>% dplyr::arrange(-logFC)

    topgenes[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    DEG_results[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe
  }
  
  ##FCG model 
  for(k in 1:length(resultlist_fcg)) {
 
    currentframe <- resultlist_fcg[[k]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    #currentframe <- currentframe %>% dplyr::filter(adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcomparison <- names(resultlist_fcg[k])
    currentmodel <- "FCG"
    currentframe$cell <- currentcell
    currentframe$model <- currentmodel
    currentframe$comparison <- currentcomparison
    currentframe <- currentframe %>% dplyr::arrange(-logFC)

    topgenes[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    DEG_results[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe
  }
  
  ##full model 
  for(k in 1:length(resultlist_full)) {
 
    currentframe <- resultlist_full[[k]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    #currentframe <- currentframe %>% dplyr::filter(adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcomparison <- names(resultlist_full[k])
    currentmodel <- "Full"
    currentframe$cell <- currentcell
    currentframe$model <- currentmodel
    currentframe$comparison <- currentcomparison
    currentframe <- currentframe %>% dplyr::arrange(-logFC)

    topgenes[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    DEG_results[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe
  }
  
  ##full continuous model 
  for(k in 1:length(resultlist_full_cont)) {
 
    currentframe <- resultlist_full_cont[[k]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    #currentframe <- currentframe %>% dplyr::filter(adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcomparison <- names(resultlist_full_cont[k])
    currentmodel <- "Full_continuous"
    currentframe$cell <- currentcell
    currentframe$model <- currentmodel
    currentframe$comparison <- currentcomparison
    currentframe <- currentframe %>% dplyr::arrange(-logFC)

    topgenes[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    DEG_results[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe
  }
  
  ##full quadratic model 
  for(k in 1:length(resultlist_full_quad)) {
 
    currentframe <- resultlist_full_quad[[k]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    #currentframe <- currentframe %>% dplyr::filter(adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcomparison <- names(resultlist_full_quad[k])
    currentmodel <- "Full_quadratic"
    currentframe$cell <- currentcell
    currentframe$model <- currentmodel
    currentframe$comparison <- currentcomparison
    currentframe <- currentframe %>% dplyr::arrange(-logFC)

    topgenes[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    DEG_results[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe
  }
  
  ##full aneuploidy model 
  for(k in 1:length(resultlist_add)) {
 
    currentframe <- resultlist_add[[k]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    #currentframe <- currentframe %>% dplyr::filter(adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcomparison <- names(resultlist_add[k])
    currentmodel <- "Full_aneuploidy"
    currentframe$cell <- currentcell
    currentframe$model <- currentmodel
    currentframe$comparison <- currentcomparison
    currentframe <- currentframe %>% dplyr::arrange(-logFC)

    topgenes[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    DEG_results[[currentmodel]][[currentcomparison]][[currentcell]] <- currentframe
  }
}

save(topgenes, DEG_results, file = "Saves/DEG_results2.Rda")
```

```{r}
##merge DEG results in 1 df 

#test <- unlist(DEG_results, use.names = TRUE, recursive = FALSE) %>% unlist(use.names=TRUE, recursive = FALSE)
#test <- map2(test, names(test), ~ mutate(.x, model.condition.cell = .y))
#final <- bind_rows(test)

load(file = "Saves/DEG_results2.Rda")

DEG_df <- do.call(c, unlist(DEG_results, recursive=FALSE)) %>% bind_rows()
DEG_df$model <- factor(DEG_df$model, levels = c("Normative", "FCG", "Full", "Full_aneuploidy", "Full_continuous", "Full_quadratic"))

gene_annotation2 <- gene_annotation[!duplicated(gene_annotation$mgi_symbol), c(1:3, 7:9)]
DEG_df_annotated <- left_join(DEG_df, gene_annotation2, by = c("gene"="mgi_symbol"))
DEG_df_annotated$chromosome_name <- factor(DEG_df_annotated$chromosome_name, levels = c((1:19),"X","Y","MT"))
```

```{r}
##elbow plots (all cell types)
DEG_df_filtered <- DEG_df %>% dplyr::filter(adj.P.Val < 0.05)

##FCG 
gene_list <- DEG_df_filtered %>% dplyr::filter(model %in% c("Normative", "FCG")) %>% dplyr::select(gene, comparison)
gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame()
gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)

set_vars <- c("GonadFemale:SexChromXX", "XX_vs_XY", "Ovary_vs_Testis", "XXF_vs_XYM")
plot <- UpSetR::upset(data = gene_list, sets = set_vars, keep.order = TRUE)
plot

pdf("Plots/DEG/elbow/FCG.pdf",width  = 10, height = 5)
print(plot)
dev.off()

##Full 
gene_list <- DEG_df_filtered %>% dplyr::filter(model %in% c("Normative", "Full")) %>% dplyr::select(gene, comparison)
gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame()
gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)

set_vars <- c("GonadFemale:Ydose2", "GonadFemale:Ydose1", "GonadFemale:Xdose2", "Ydose2_vs_Ydose1", "Ydose2_vs_Ydose0", "Ydose1_vs_Ydose0", "Xdose2_vs_Xdose1", "Ovary_vs_Testis", "XXF_vs_XYM")
plot <- UpSetR::upset(data = gene_list, sets = set_vars, keep.order = TRUE)
plot

pdf("Plots/DEG/elbow/Full.pdf",width  = 10, height = 5)
print(plot)
dev.off()

##Full aneuploidy
gene_list <- DEG_df_filtered %>% dplyr::filter(model %in% c("Normative", "Full_aneuploidy")) %>% dplyr::select(gene, comparison)
gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% data.frame(check.names = FALSE)
gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)

set_vars <- c("GonadFemale:Aneuploidy", "GonadFemale:base_SexChromXX", "Aneuploidy", "XX_vs_XY", "Ovary_vs_Testis", "XXF_vs_XYM")
plot <- UpSetR::upset(data = gene_list, sets = set_vars, keep.order = TRUE)
plot

pdf("Plots/DEG/elbow/Full_aneuploidy.pdf",width  = 10, height = 5)
print(plot)
dev.off()

##Full continuous
gene_list <- DEG_df_filtered %>% dplyr::filter(model %in% c("Normative", "Full_continuous")) %>% dplyr::select(gene, comparison)
gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame(check.names = FALSE)
gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)

set_vars <- c("Xdose:Ydose", "GonadFemale:Ydose", "GonadFemale:Xdose", "Ydose", "Xdose", "Ovary_vs_Testis", "XXF_vs_XYM")
plot <- UpSetR::upset(data = gene_list, sets = set_vars, keep.order = TRUE)
plot

pdf("Plots/DEG/elbow/Full_continuous.pdf",width  = 10, height = 5)
print(plot)
dev.off()

##Full quadratic
gene_list <- DEG_df_filtered %>% dplyr::filter(model %in% c("Normative", "Full_quadratic")) %>% dplyr::select(gene, comparison)
gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame(check.names=FALSe)
gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)

set_vars <- c("Ydose^2", "Ydose", "Xdose", "Ovary_vs_Testis", "XXF_vs_XYM")
plot <- UpSetR::upset(data = gene_list, sets = set_vars, keep.order = TRUE)
plot

pdf("Plots/DEG/elbow/Full_quadratic.pdf",width  = 10, height = 5)
print(plot)
dev.off()
```

```{r}
##elbow plots (individual cell types)
DEG_df_filtered <- DEG_df %>% dplyr::filter(adj.P.Val < 0.05) %>% dplyr::mutate(comparison = factor(comparison))

##FCG 
plot_list <- list()
for (i in 1:length(allcells)) {
 
  currentcell <- as.character(allcells[i])
  gene_list <- DEG_df_filtered %>% dplyr::filter(model %in% c("Normative", "FCG") & cell == currentcell) %>% dplyr::select(gene, comparison)
  gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
  gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame(check.names = FALSE)
  gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)
  
  set_vars <- c("GonadFemale:SexChromXX", "XX_vs_XY", "Ovary_vs_Testis", "XXF_vs_XYM")
  gene_list <- gene_list %>% dplyr::select(Var1, all_of(set_vars))
  
  plot <- ComplexUpset::upset(gene_list, set_vars, sort_sets = FALSE, sort_intersections = FALSE, name = currentcell,
                              base_annotations = list(
                                'Intersection size'=(intersection_size(text=list(size=5),
                                                                       width=0.4,
                                                                       text_colors=c(on_background='black',on_bar='black'),
                                                                                     bar_number_threshold=1.0)+ 
                                                                         theme(panel.grid.major = element_blank(),
                                                                               panel.grid.minor = element_blank()))))
  plot_list[[currentcell]] <- plot
}

pdf("Plots/DEG/elbow_celltype/FCG.pdf", width = 40, height = 15)
do.call(ggarrange, c(plot_list, list(ncol = 4, nrow = 2)))
dev.off()

##Full
plot_list <- list()
for (i in 1:length(allcells)) {
 
  currentcell <- as.character(allcells[i])
  gene_list <- DEG_df_filtered %>% dplyr::filter(model %in% c("Normative", "Full") & cell == currentcell) %>% dplyr::select(gene, comparison)
  gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
  gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame(check.names = FALSE)
  gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)
  
  set_vars <- c("GonadFemale:Ydose2", "GonadFemale:Ydose1", "GonadFemale:Xdose2", "Ydose2_vs_Ydose1", "Ydose2_vs_Ydose0", "Ydose1_vs_Ydose0", "Xdose2_vs_Xdose1", "Ovary_vs_Testis", "XXF_vs_XYM")
  gene_list <- gene_list %>% dplyr::select(Var1, any_of(set_vars))
  #gene_list <- gene_list[, c("TRUE", colnames(gene_list) %in% set_vars)]
  
  plot <- ComplexUpset::upset(gene_list, names(gene_list)[-1], sort_sets = FALSE, sort_intersections = FALSE, name = currentcell,
                              themes=upset_modify_themes(list('intersections_matrix'=theme(axis.text.y=element_text(size=20)))),
                              base_annotations = list(
                                'Intersection size'=(intersection_size(text=list(size=3),
                                                                       width=0.4,
                                                                       text_colors=c(on_background='black',on_bar='black'),
                                                                                     bar_number_threshold=1.0)+ 
                                                                         theme(panel.grid.major = element_blank(),
                                                                               panel.grid.minor = element_blank()))))
  plot_list[[currentcell]] <- plot
}

pdf("Plots/DEG/elbow_celltype/Full.pdf", width  = 70, height = 15)
do.call(ggarrange, c(plot_list, list(ncol = 4, nrow = 2)))
dev.off()

##Full aneuploidy
plot_list <- list()
for (i in 1:length(allcells)) {
 
  currentcell <- as.character(allcells[i])
  gene_list <- DEG_df_filtered %>% dplyr::filter(model %in% c("Normative", "Full_aneuploidy") & cell == currentcell) %>% dplyr::select(gene, comparison)
  gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
  gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame(check.names = FALSE)
  gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)
  
  set_vars <- c("GonadFemale:Aneuploidy", "GonadFemale:base_SexChromXX", "Aneuploidy", "XX_vs_XY", "Ovary_vs_Testis", "XXF_vs_XYM")
  gene_list <- gene_list %>% dplyr::select(Var1, all_of(set_vars))
  
  plot <- ComplexUpset::upset(gene_list, set_vars, sort_sets = FALSE, sort_intersections = FALSE, name = currentcell,
                              base_annotations = list(
                                'Intersection size'=(intersection_size(text=list(size=3),
                                                                       width=0.4,
                                                                       text_colors=c(on_background='black',on_bar='black'),
                                                                                     bar_number_threshold=1.0)+ 
                                                                         theme(panel.grid.major = element_blank(),
                                                                               panel.grid.minor = element_blank()))))
  plot_list[[currentcell]] <- plot
}

pdf("Plots/DEG/elbow_celltype/Full_aneuploidy.pdf", width  = 60, height = 15)
do.call(ggarrange, c(plot_list, list(ncol = 4, nrow = 2)))
dev.off()

##Full continuous
plot_list <- list()
for (i in 1:length(allcells)) {
 
  currentcell <- as.character(allcells[i])
  gene_list <- DEG_df_filtered %>% dplyr::filter(model %in% c("Normative", "Full_continuous") & cell == currentcell) %>% dplyr::select(gene, comparison)
  gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
  gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame(check.names = FALSE)
  gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)
  
  set_vars <- c("Xdose:Ydose", "GonadFemale:Ydose", "GonadFemale:Xdose", "Ydose", "Xdose", "Ovary_vs_Testis", "XXF_vs_XYM")
  gene_list <- gene_list %>% dplyr::select(Var1, all_of(set_vars))
  
  plot <- ComplexUpset::upset(gene_list, set_vars, sort_sets = FALSE, sort_intersections = FALSE, name = currentcell,
                              base_annotations = list(
                                'Intersection size'=(intersection_size(text=list(size=3),
                                                                       width=0.4,
                                                                       text_colors=c(on_background='black',on_bar='black'),
                                                                                     bar_number_threshold=1.0)+ 
                                                                         theme(panel.grid.major = element_blank(),
                                                                               panel.grid.minor = element_blank()))))
  plot_list[[currentcell]] <- plot
}

pdf("Plots/DEG/elbow_celltype/Full_continuous.pdf", width  = 60, height = 15)
do.call(ggarrange, c(plot_list, list(ncol = 4, nrow = 2)))
dev.off()


##Full quadratic
plot_list <- list()
for (i in 1:length(allcells)) {
 
  currentcell <- as.character(allcells[i])
  gene_list <- DEG_df_filtered %>% dplyr::filter(model %in% c("Normative", "Full_quadratic") & cell == currentcell) %>% dplyr::select(gene, comparison)
  gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
  gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame(check.names = FALSE)
  gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)
  
  set_vars <- c("Ydose^2", "Ydose", "Xdose", "Ovary_vs_Testis", "XXF_vs_XYM")
  gene_list <- gene_list %>% dplyr::select(Var1, all_of(set_vars))
  
  plot <- ComplexUpset::upset(gene_list, set_vars, sort_sets = FALSE, sort_intersections = FALSE, name = currentcell,
                              base_annotations = list(
                                'Intersection size'=(intersection_size(text=list(size=3),
                                                                       width=0.4,
                                                                       text_colors=c(on_background='black',on_bar='black'),
                                                                                     bar_number_threshold=1.0)+ 
                                                                         theme(panel.grid.major = element_blank(),
                                                                               panel.grid.minor = element_blank()))))
  plot_list[[currentcell]] <- plot
}

pdf("Plots/DEG/elbow_celltype/Full_quadratic.pdf", width  = 50, height = 15)
do.call(ggarrange, c(plot_list, list(ncol = 4, nrow = 2)))
dev.off()

```

```{r}
##elbow plot - old models 
##XY analysis 
gene_list <- DEG_df %>% dplyr::filter(model %in% c("Normative", "XY_males")) %>% dplyr::select(gene, comparison)
gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame()
gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)

set_vars <- c("XYY_vs_XY", "XXY_vs_XY", "XXF_vs_XYM")
plot <- UpSetR::upset(data = gene_list, sets = set_vars, keep.order = TRUE)
plot

pdf("Plots/DEG/elbow/XY_males.pdf",width  = 10, height = 5)
print(plot)
dev.off()

gene_list <- DEG_df %>% dplyr::filter(model %in% c("Normative", "XY_females")) %>% dplyr::select(gene, comparison)
gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame()
gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)

set_vars <- c("XYY_vs_XY", "XXY_vs_XY", "XXF_vs_XYM")
plot <- UpSetR::upset(data = gene_list, sets = set_vars, keep.order = TRUE)
plot

pdf("Plots/DEG/elbow/XY_females.pdf",width  = 10, height = 5)
print(plot)
dev.off()

gene_list <- DEG_df %>% dplyr::filter(model %in% c("Normative", "XY_combined")) %>% dplyr::select(gene, comparison)
gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame()
gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)

set_vars <- c("XYY_vs_XY", "XXY_vs_XY", "XXF_vs_XYM")
plot <- UpSetR::upset(data = gene_list, sets = set_vars, keep.order = TRUE)
plot

pdf("Plots/DEG/elbow/XY_combined.pdf",width  = 10, height = 5)
print(plot)
dev.off()

##Full
gene_list <- DEG_df %>% dplyr::filter(model %in% c("Normative", "Full")) %>% dplyr::select(gene, comparison)
gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame()
gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)

set_vars <- c("GonadFemale:Ydose2", "GonadFemale:Ydose1", "GonadFemale:Xdose2", "Ydose2_vs_Ydose1", "Ydose2_vs_Ydose0", "Ydose1_vs_Ydose0", "Xdose2_vs_Xdose1", "Ovary_vs_Testis", "XXF_vs_XYM")
plot <- UpSetR::upset(data = gene_list, sets = set_vars, keep.order = TRUE)
plot

pdf("Plots/DEG/elbow/Full.pdf",width  = 10, height = 5)
print(plot)
dev.off()

##AdditionalX
gene_list <- DEG_df %>% dplyr::filter(model %in% c("Normative", "additionalX")) %>% dplyr::select(gene, comparison)
gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame()
gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)

set_vars <- c("addX1", "XXF_vs_XYM")
plot <- UpSetR::upset(data = gene_list, sets = set_vars, keep.order = TRUE)
plot

pdf("Plots/DEG/elbow/addX.pdf",width  = 10, height = 5)
print(plot)
dev.off()

gene_list <- DEG_df %>% dplyr::filter(model %in% c("Normative", "additionalX_continuous")) %>% dplyr::select(gene, comparison)
gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame()
gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)

set_vars <- c("addX", "XXF_vs_XYM")
plot <- UpSetR::upset(data = gene_list, sets = set_vars, keep.order = TRUE)
plot

pdf("Plots/DEG/elbow/addX_continuous.pdf",width  = 10, height = 5)
print(plot)
dev.off()

##AdditionalY
gene_list <- DEG_df %>% dplyr::filter(model %in% c("Normative", "additionalY")) %>% dplyr::select(gene, comparison)
gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame()
gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)

set_vars <- c("addY2", "addY1", "XXF_vs_XYM")
plot <- UpSetR::upset(data = gene_list, sets = set_vars, keep.order = TRUE)
plot

pdf("Plots/DEG/elbow/addY.pdf",width  = 10, height = 5)
print(plot)
dev.off()

gene_list <- DEG_df %>% dplyr::filter(model %in% c("Normative", "additionalY_continuous")) %>% dplyr::select(gene, comparison)
gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame()
gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)

set_vars <- c("addY", "XXF_vs_XYM")
plot <- UpSetR::upset(data = gene_list, sets = set_vars, keep.order = TRUE)
plot

pdf("Plots/DEG/elbow/addY_continuous.pdf",width  = 10, height = 5)
print(plot)
dev.off()
```

```{r}
##create barplot of DEG counts by chromosome and gene category by model and overall  

##FCG
test_df <- DEG_df_annotated %>% dplyr::filter(model %in% c("Normative", "FCG"))

barplot_df <- data.frame(table(test_df$comparison, test_df$chromosome_name))
barplot_df$Var1 <- factor(barplot_df$Var1, levels = c("XXF_vs_XYM", "Ovary_vs_Testis", "XX_vs_XY", "Xdose", "GonadFemale:SexChromXX"))

plot <- ggplot(barplot_df, aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Var2) + 
  scale_x_discrete(labels = NULL, breaks = NULL) + labs(x = NULL) +
  guides(fill=guide_legend(title="comparison")) +
  theme(legend.title = element_text(size=16),
        legend.text = element_text(size=12),
        strip.text.x = element_text(size = 12),
        axis.text.x = element_text(size = 12))

pdf("Plots/DEG/bar/FCG_chromosome.pdf",width  = 13, height = 10)
print(plot)
dev.off()

barplot_df <- data.frame(table(test_df$comparison, test_df$Class))
barplot_df$Var1 <- factor(barplot_df$Var1, levels = c("XXF_vs_XYM", "Ovary_vs_Testis", "XX_vs_XY", "Xdose", "GonadFemale:SexChromXX"))

plot <- ggplot(barplot_df, aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Var2) + 
  scale_x_discrete(labels = NULL, breaks = NULL) + labs(x = NULL) +
  theme(legend.title = element_text(size=16),
        legend.text = element_text(size=12),
        strip.text.x = element_text(size = 12),
        axis.text.x = element_text(size = 12))

pdf("Plots/DEG/bar/FCG_class.pdf",width  = 12, height = 5)
print(plot)
dev.off()

barplot_df <- data.frame(table(test_df$comparison, test_df$PAR))
barplot_df$Var1 <- factor(barplot_df$Var1, levels = c("XXF_vs_XYM", "Ovary_vs_Testis", "XX_vs_XY", "Xdose", "GonadFemale:SexChromXX"))

plot <- ggplot(barplot_df, aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Var2) + 
  scale_x_discrete(labels = NULL, breaks = NULL) + labs(x = NULL) +
  theme(legend.title = element_text(size=16),
        legend.text = element_text(size=12),
        strip.text.x = element_text(size = 12),
        axis.text.x = element_text(size = 12))

pdf("Plots/DEG/bar/FCG_par.pdf",width  = 12, height = 5)
print(plot)
dev.off()

test_df <- test_df %>% dplyr::filter(chromosome_name == "X")
barplot_df <- data.frame(table(test_df$comparison, test_df$Escapee))
barplot_df$Var1 <- factor(barplot_df$Var1, levels = c("XXF_vs_XYM", "Ovary_vs_Testis", "XX_vs_XY", "Xdose", "GonadFemale:SexChromXX"))

plot <- ggplot(barplot_df, aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Var2) + 
  scale_x_discrete(labels = NULL, breaks = NULL) + labs(x = NULL) +
  theme(legend.title = element_text(size=16),
        legend.text = element_text(size=12),
        strip.text.x = element_text(size = 12),
        axis.text.x = element_text(size = 12))

pdf("Plots/DEG/bar/FCG_escape.pdf",width  = 12, height = 5)
print(plot)
dev.off()

##XY analysis
test_df <- DEG_df_annotated %>% 
  mutate(comparison = ifelse(model == c("XY_males"), paste(DEG_df_annotated$comparison, "male", sep = "_"), 
                             ifelse(model == c("XY_females"), paste(DEG_df_annotated$comparison, "female", sep = "_"), 
                                    ifelse(model == c("XY_combined"), paste(DEG_df_annotated$comparison, "combined", sep = "_"),
                                           DEG_df_annotated$comparison))))
                             
test_df <- test_df %>% dplyr::filter(model %in% c("Normative", "XY_males", "XY_females", "XY_combined"))

barplot_df <- data.frame(table(test_df$comparison, test_df$chromosome_name))
barplot_df$Var1 <- factor(barplot_df$Var1, levels = c("XXF_vs_XYM", "XXY_vs_XY_male", "XXY_vs_XY_female", "XXY_vs_XY_combined", "XYY_vs_XY_male", "XYY_vs_XY_female", "XYY_vs_XY_combined"))

plot <- ggplot(barplot_df, aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Var2) + 
  scale_x_discrete(labels = NULL, breaks = NULL) + labs(x = NULL)  +
  theme(legend.title = element_text(size=16),
        legend.text = element_text(size=12),
        strip.text.x = element_text(size = 12),
        axis.text.x = element_text(size = 12))

pdf("Plots/DEG/bar/XY_chromosome.pdf",width  = 13, height = 10)
print(plot)
dev.off()

barplot_df <- data.frame(table(test_df$comparison, test_df$Class))
barplot_df$Var1 <- factor(barplot_df$Var1, levels = c("XXF_vs_XYM", "XXY_vs_XY_male", "XXY_vs_XY_female", "XXY_vs_XY_combined", "XYY_vs_XY_male", "XYY_vs_XY_female", "XYY_vs_XY_combined"))

plot <- ggplot(barplot_df, aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Var2) + 
  scale_x_discrete(labels = NULL, breaks = NULL) + labs(x = NULL) +
  theme(legend.title = element_text(size=16),
        legend.text = element_text(size=12),
        strip.text.x = element_text(size = 12),
        axis.text.x = element_text(size = 12))

pdf("Plots/DEG/bar/XY_class.pdf",width  = 12, height = 5)
print(plot)
dev.off()

barplot_df <- data.frame(table(test_df$comparison, test_df$PAR))
barplot_df$Var1 <- factor(barplot_df$Var1, levels = c("XXF_vs_XYM", "XXY_vs_XY_male", "XXY_vs_XY_female", "XXY_vs_XY_combined", "XYY_vs_XY_male", "XYY_vs_XY_female", "XYY_vs_XY_combined"))

plot <- ggplot(barplot_df, aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Var2) + 
  scale_x_discrete(labels = NULL, breaks = NULL) + labs(x = NULL) +
  theme(legend.title = element_text(size=16),
        legend.text = element_text(size=12),
        strip.text.x = element_text(size = 12),
        axis.text.x = element_text(size = 12))

pdf("Plots/DEG/bar/XY_par.pdf",width  = 12, height = 5)
print(plot)
dev.off()

test_df <- test_df %>% dplyr::filter(chromosome_name == "X")
barplot_df <- data.frame(table(test_df$comparison, test_df$Escapee))
barplot_df$Var1 <- factor(barplot_df$Var1, levels = c("XXF_vs_XYM", "XXY_vs_XY_male", "XXY_vs_XY_female", "XXY_vs_XY_combined", "XYY_vs_XY_male", "XYY_vs_XY_female", "XYY_vs_XY_combined"))

plot <- ggplot(barplot_df, aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Var2) + 
  scale_x_discrete(labels = NULL, breaks = NULL) + labs(x = NULL) +
  theme(legend.title = element_text(size=16),
        legend.text = element_text(size=12),
        strip.text.x = element_text(size = 12),
        axis.text.x = element_text(size = 12))

pdf("Plots/DEG/bar/XY_escape.pdf",width  = 12, height = 5)
print(plot)
dev.off()

##Full analysis
test_df <- DEG_df_annotated %>% dplyr::filter(model %in% c("Normative", "Full"))

barplot_df <- data.frame(table(test_df$comparison, test_df$chromosome_name))
levels(barplot_df$Var1) <- c("XXF_vs_XYM", "Ovary_vs_Testis", "Xdose2_vs_Xdose1", "Ydose1_vs_Ydose0", "Ydose2-Ydose0", "Ydose2_vs_Ydose1", "GonadFemale:Xdose2", "GonadFemale:Ydose1", "GonadFemale:Ydose2")

plot <- ggplot(barplot_df, aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Var2) + 
  scale_x_discrete(labels = NULL, breaks = NULL) + labs(x = NULL)  +
  theme(legend.title = element_text(size=16),
        legend.text = element_text(size=12),
        strip.text.x = element_text(size = 12),
        axis.text.x = element_text(size = 12))

pdf("Plots/DEG/bar/full_chromosome.pdf",width  = 13, height = 10)
print(plot)
dev.off()

barplot_df <- data.frame(table(test_df$comparison, test_df$Class))
levels(barplot_df$Var1) <- c("XXF_vs_XYM", "Ovary_vs_Testis", "Xdose2_vs_Xdose1", "Ydose1_vs_Ydose0", "Ydose2-Ydose0", "Ydose2_vs_Ydose1", "GonadFemale:Xdose2", "GonadFemale:Ydose1", "GonadFemale:Ydose2")

plot <- ggplot(barplot_df, aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Var2) + 
  scale_x_discrete(labels = NULL, breaks = NULL) + labs(x = NULL) +
  theme(legend.title = element_text(size=16),
        legend.text = element_text(size=12),
        strip.text.x = element_text(size = 12),
        axis.text.x = element_text(size = 12))

pdf("Plots/DEG/bar/full_class.pdf",width  = 12, height = 5)
print(plot)
dev.off()

barplot_df <- data.frame(table(test_df$comparison, test_df$PAR))
levels(barplot_df$Var1) <- c("XXF_vs_XYM", "Ovary_vs_Testis", "Xdose2_vs_Xdose1", "Ydose1_vs_Ydose0", "Ydose2-Ydose0", "Ydose2_vs_Ydose1", "GonadFemale:Xdose2", "GonadFemale:Ydose1", "GonadFemale:Ydose2")

plot <- ggplot(barplot_df, aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Var2) + 
  scale_x_discrete(labels = NULL, breaks = NULL) + labs(x = NULL) +
  theme(legend.title = element_text(size=16),
        legend.text = element_text(size=12),
        strip.text.x = element_text(size = 12),
        axis.text.x = element_text(size = 12))

pdf("Plots/DEG/bar/full_par.pdf",width  = 12, height = 5)
print(plot)
dev.off()

test_df <- test_df %>% dplyr::filter(chromosome_name == "X")
barplot_df <- data.frame(table(test_df$comparison, test_df$Escapee))
levels(barplot_df$Var1) <- c("XXF_vs_XYM", "Ovary_vs_Testis", "Xdose2_vs_Xdose1", "Ydose1_vs_Ydose0", "Ydose2-Ydose0", "Ydose2_vs_Ydose1", "GonadFemale:Xdose2", "GonadFemale:Ydose1", "GonadFemale:Ydose2")

plot <- ggplot(barplot_df, aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Var2) + 
  scale_x_discrete(labels = NULL, breaks = NULL) + labs(x = NULL) +
  theme(legend.title = element_text(size=16),
        legend.text = element_text(size=12),
        strip.text.x = element_text(size = 12),
        axis.text.x = element_text(size = 12))

pdf("Plots/DEG/bar/full_escape.pdf",width  = 12, height = 5)
print(plot)
dev.off()

##Full continuous analysis 
test_df <- DEG_df_annotated %>% dplyr::filter(model %in% c("Normative", "Full_continuous"))

barplot_df <- data.frame(table(test_df$comparison, test_df$chromosome_name))
levels(barplot_df$Var1) <- c("XXF_vs_XYM", "Ovary_vs_Testis", "Xdose", "Ydose", "GonadFemale:Xdose", "GonadFemale:Ydose", "Xdose:Ydose")

plot <- ggplot(barplot_df, aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Var2) + 
  scale_x_discrete(labels = NULL, breaks = NULL) + labs(x = NULL)  +
  theme(legend.title = element_text(size=16),
        legend.text = element_text(size=12),
        strip.text.x = element_text(size = 12),
        axis.text.x = element_text(size = 12))

pdf("Plots/DEG/bar/full_cont_chromosome.pdf",width  = 13, height = 10)
print(plot)
dev.off()

barplot_df <- data.frame(table(test_df$comparison, test_df$Class))
levels(barplot_df$Var1) <- c("XXF_vs_XYM", "Ovary_vs_Testis", "Xdose", "Ydose", "GonadFemale:Xdose", "GonadFemale:Ydose", "Xdose:Ydose")

plot <- ggplot(barplot_df, aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Var2) + 
  scale_x_discrete(labels = NULL, breaks = NULL) + labs(x = NULL) +
  theme(legend.title = element_text(size=16),
        legend.text = element_text(size=12),
        strip.text.x = element_text(size = 12),
        axis.text.x = element_text(size = 12))

pdf("Plots/DEG/bar/full_cont_class.pdf",width  = 12, height = 5)
print(plot)
dev.off()

barplot_df <- data.frame(table(test_df$comparison, test_df$PAR))
levels(barplot_df$Var1) <- c("XXF_vs_XYM", "Ovary_vs_Testis", "Xdose", "Ydose", "GonadFemale:Xdose", "GonadFemale:Ydose", "Xdose:Ydose")

plot <- ggplot(barplot_df, aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Var2) + 
  scale_x_discrete(labels = NULL, breaks = NULL) + labs(x = NULL) +
  theme(legend.title = element_text(size=16),
        legend.text = element_text(size=12),
        strip.text.x = element_text(size = 12),
        axis.text.x = element_text(size = 12))

pdf("Plots/DEG/bar/full_cont_par.pdf",width  = 12, height = 5)
print(plot)
dev.off()

test_df <- test_df %>% dplyr::filter(chromosome_name == "X")
barplot_df <- data.frame(table(test_df$comparison, test_df$Escapee))
levels(barplot_df$Var1) <- c("XXF_vs_XYM", "Ovary_vs_Testis", "Xdose", "Ydose", "GonadFemale:Xdose", "GonadFemale:Ydose", "Xdose:Ydose")

plot <- ggplot(barplot_df, aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Var2) + 
  scale_x_discrete(labels = NULL, breaks = NULL) + labs(x = NULL) +
  theme(legend.title = element_text(size=16),
        legend.text = element_text(size=12),
        strip.text.x = element_text(size = 12),
        axis.text.x = element_text(size = 12))

pdf("Plots/DEG/bar/full_cont_escape.pdf",width  = 12, height = 5)
print(plot)
dev.off()
```

```{r}
##p-value dotplots 
ygenes <- DEG_df_annotated %>% dplyr::filter(chromosome_name == "Y") %>% dplyr::arrange(adj.P.Val) %>% dplyr::select(gene) %>% distinct()
xgenes <- DEG_df_annotated %>% dplyr::filter(chromosome_name == "X" & Escapee == "Non-escapee") %>% dplyr::arrange(adj.P.Val) %>% dplyr::select(gene) %>% distinct() %>% head(20)
xgenes_escape <- DEG_df_annotated %>% dplyr::filter(Escapee == "Escapee") %>% dplyr::select(gene) %>% distinct()
genes_to_plot <- c(xgenes_escape, xgenes, ygenes) %>% unlist(use.names = FALSE)

DEG_df_annotated <- DEG_df_annotated %>% dplyr::filter(adj.P.Val < 0.05)

allcells <- unique(seurat$cell.type)
allcells <- allcells[!allcells == "Unknown"]

for (i in 1:length(allcells)) {
  plot_list <- list()
  currentcell <- allcells[i]
  
  #Normative
  plot_df <- DEG_df_annotated %>% 
    dplyr::filter(model == "Normative" & cell == currentcell) %>%
    dplyr::filter(gene %in% genes_to_plot) %>%
    dplyr::arrange(match(gene, genes_to_plot)) %>%
    mutate(gene = factor(gene, levels = rev(genes_to_plot)))
  
  plot_df$comparison <- factor(plot_df$comparison, levels = c("XXF_vs_XYM"))
  
  plot <- ggplot(plot_df, aes(y = gene, x = comparison)) +
    geom_tile(fill = "white") +  xlab("")+ylab("") +
    geom_point(aes(colour = logFC, size = adj.P.Val))  +  
    scale_color_gradientn(limits = c(-1.5,1.5), colours  = colorRampPalette(rev(brewer.pal(3,"RdBu")))(10)) +
    theme_bw()+theme(axis.text.y = element_text( size = 20),
                     axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                     strip.text.x = element_text(size = 25),
                     plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                     plot.title = element_text(size = 25)) +
    scale_size(range = c(5, 10))
  
  plot_list[["Normative"]] <- plot
  
  #pdf(paste0("Plots/DEG/dot/Normative_", currentcell, ".pdf"), width  = 12, height = 25)
  #print(plot)
  #dev.off()
  
  #FCG
  plot_df <- DEG_df_annotated %>% 
    dplyr::filter(model == "FCG" & cell == currentcell) %>%
    dplyr::filter(gene %in% genes_to_plot) %>%
    dplyr::arrange(match(gene, genes_to_plot)) %>%
    mutate(gene = factor(gene, levels = rev(genes_to_plot)))
  
  plot_df$comparison <- factor(plot_df$comparison, levels = c("Ovary_vs_Testis", "XX_vs_XY", "GonadFemale:SexChromXX"))
  
  plot <- ggplot(plot_df, aes(y = gene, x = comparison)) +
    geom_tile(fill = "white") +  xlab("")+ylab("") +
    geom_point(aes(colour = logFC, size = adj.P.Val))  +  
    scale_color_gradientn(limits = c(-1.5,1.5), colours  = colorRampPalette(rev(brewer.pal(3,"RdBu")))(10)) +
    theme_bw()+theme(axis.text.y = element_text( size = 20),
                     axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                     strip.text.x = element_text(size = 25),
                     plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                     plot.title = element_text(size = 25)) +
    scale_size(range = c(5, 10))
  
  plot_list[["FCG"]] <- plot
  
  #pdf(paste0("Plots/DEG/dot/FCG_", currentcell, ".pdf"), width  = 12, height = 25)
  #print(plot)
  #dev.off()
  
  #Full
  plot_df <- DEG_df_annotated %>% 
    dplyr::filter(model == "Full" & cell == currentcell) %>%
    dplyr::filter(gene %in% genes_to_plot) %>%
    dplyr::arrange(match(gene, genes_to_plot)) %>%
    mutate(gene = factor(gene, levels = rev(genes_to_plot)))
  
  plot_df$comparison = factor(plot_df$comparison, levels = c("Ovary_vs_Testis", "Xdose2_vs_Xdose1", "Ydose1_vs_Ydose0", "Ydose2_vs_Ydose0", "Ydose2_vs_Ydose1", "GonadFemale:Xdose2", "GonadFemale:Ydose1", "GonadFemale:Ydose2")) 
  
  plot <- ggplot(plot_df, aes(y = gene, x = comparison)) +
    geom_tile(fill = "white") +  xlab("")+ylab("") +
    geom_point(aes(colour = logFC, size = adj.P.Val))  +  
    scale_color_gradientn(limits = c(-1.5,1.5), colours  = colorRampPalette(rev(brewer.pal(3,"RdBu")))(10)) +
    theme_bw()+theme(axis.text.y = element_text( size = 20),
                     axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                     strip.text.x = element_text(size = 25),
                     plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                     plot.title = element_text(size = 25)) +
    scale_size(range = c(5, 10))
  
  plot_list[["Full"]] <- plot
  
  #pdf(paste0("Plots/DEG/dot/Full_", currentcell, ".pdf"), width  = 12, height = 25)
  #print(plot)
  #dev.off()
  
  #Full_aneuploidy
  plot_df <- DEG_df_annotated %>% 
    dplyr::filter(model == "Full_aneuploidy" & cell == currentcell) %>%
    dplyr::filter(gene %in% genes_to_plot) %>%
    dplyr::arrange(match(gene, genes_to_plot)) %>%
    mutate(gene = factor(gene, levels = rev(genes_to_plot)))
  
  plot_df$comparison <- factor(plot_df$comparison, levels = c("Ovary_vs_Testis", "XX_vs_XY", "Aneuploidy", "GonadFemale:base_SexChromXX", "GonadFemale:Aneuploidy"))
  
  plot <- ggplot(plot_df, aes(y = gene, x = comparison)) +
    geom_tile(fill = "white") +  xlab("")+ylab("") +
    geom_point(aes(colour = logFC, size = adj.P.Val))  +  
    scale_color_gradientn(limits = c(-1.5,1.5), colours  = colorRampPalette(rev(brewer.pal(3,"RdBu")))(10)) +
    theme_bw()+theme(axis.text.y = element_text( size = 20),
                     axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                     strip.text.x = element_text(size = 25),
                     plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                     plot.title = element_text(size = 25)) +
    scale_size(range = c(5, 10))
  
  plot_list[["Full_aneuploidy"]] <- plot
  
  #pdf(paste0("Plots/DEG/dot/Full_add_", currentcell, ".pdf"), width  = 12, height = 25)
  #print(plot)
  #dev.off()
  
  #Full_continuous
  plot_df <- DEG_df_annotated %>% 
    dplyr::filter(model == "Full_continuous" & cell == currentcell) %>%
    dplyr::filter(gene %in% genes_to_plot) %>%
    dplyr::arrange(match(gene, genes_to_plot)) %>%
    mutate(gene = factor(gene, levels = rev(genes_to_plot)))
  
  plot_df$comparison <- factor(plot_df$comparison, levels = c("Ovary_vs_Testis", "Xdose", "Ydose", "GonadFemale:Xdose", "GonadFemale:Ydose", "Xdose:Ydose"))
  
  plot <- ggplot(plot_df, aes(y = gene, x = comparison)) +
    geom_tile(fill = "white") +  xlab("")+ylab("") +
    geom_point(aes(colour = logFC, size = adj.P.Val))  +  
    scale_color_gradientn(limits = c(-1.5,1.5), colours  = colorRampPalette(rev(brewer.pal(3,"RdBu")))(10)) +
    theme_bw()+theme(axis.text.y = element_text( size = 20),
                     axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                     strip.text.x = element_text(size = 25),
                     plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                     plot.title = element_text(size = 25)) +
    scale_size(range = c(5, 10))
  
  plot_list[["Full_continuous"]] <- plot
  
  #pdf(paste0("Plots/DEG/dot/Full_cont_", currentcell, ".pdf"), width  = 12, height = 25)
  #print(plot)
  #dev.off()
  
  #Full_quadratic
  plot_df <- DEG_df_annotated %>% 
    dplyr::filter(model == "Full_quadratic" & cell == currentcell) %>%
    dplyr::filter(gene %in% genes_to_plot) %>%
    dplyr::arrange(match(gene, genes_to_plot)) %>%
    mutate(gene = factor(gene, levels = rev(genes_to_plot)))
  
  plot_df$comparison <- factor(plot_df$comparison, levels = c("Ovary_vs_Testis", "Xdose", "Ydose", "Ydose^2"))
  
  plot <- ggplot(plot_df, aes(y = gene, x = comparison)) +
    geom_tile(fill = "white") +  xlab("")+ylab("") +
    geom_point(aes(colour = logFC, size = adj.P.Val))  +  
    scale_color_gradientn(limits = c(-1.5,1.5), colours  = colorRampPalette(rev(brewer.pal(3,"RdBu")))(10)) +
    theme_bw()+theme(axis.text.y = element_text( size = 20),
                     axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                     strip.text.x = element_text(size = 25),
                     plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                     plot.title = element_text(size = 25)) +
    scale_size(range = c(5, 10))
  
  plot_list[["Full_quad"]] <- plot
  
  #pdf(paste0("Plots/DEG/dot/Full_quad_", currentcell, ".pdf"), width  = 12, height = 25)
  #print(plot)
  #dev.off()
  
  pdf(paste0("Plots/DEG/dot/", currentcell, ".pdf"), width  = 60, height = 25)
  print(ggarrange(plotlist = plot_list, labels = names(plot_list), nrow = 1))
  dev.off()
}
```

```{r}

##dotplot - old 
#Normative
plot_df <- DEG_df_annotated %>% 
  dplyr::filter(model == "Normative" & cell == "Neuron") %>%
  dplyr::filter(gene %in% genes_to_plot) %>%
  dplyr::arrange(match(gene, genes_to_plot)) %>%
  mutate(gene = factor(gene, levels = rev(genes_to_plot)))

plot_df$comparison <- factor(plot_df$comparison, levels = c("XXF_vs_XYM"))

plot <- ggplot(plot_df, aes(y = gene, x = comparison)) +
  geom_tile(fill = "white") +  xlab("")+ylab("") +
  geom_point(aes(colour = logFC, size = adj.P.Val))  +  
  scale_color_gradientn(limits = c(-1.5,1.5), colours  = colorRampPalette(rev(brewer.pal(3,"RdBu")))(10)) +
  theme_bw()+theme(axis.text.y = element_text( size = 20),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25)) +
  scale_size(range = c(5, 10))

pdf("Plots/DEG/dot/Normative.pdf", width  = 12, height = 25)
print(plot)
dev.off()

#FCG
plot_df <- DEG_df_annotated %>% 
  dplyr::filter(model == "FCG" & cell == "Neuron" & comparison %in% c("Ovary_vs_Testis", "XX_vs_XY", "Xdose", "GonadFemale:SexChromXX")) %>%
  dplyr::filter(gene %in% genes_to_plot) %>%
  dplyr::arrange(match(gene, genes_to_plot)) %>%
  mutate(gene = factor(gene, levels = rev(genes_to_plot)))

plot_df$comparison <- factor(plot_df$comparison, levels = c("Ovary_vs_Testis", "XX_vs_XY", "Xdose", "GonadFemale:SexChromXX"))

plot <- ggplot(plot_df, aes(y = gene, x = comparison)) +
  geom_tile(fill = "white") +  xlab("")+ylab("") +
  geom_point(aes(colour = logFC, size = adj.P.Val))  +  
  scale_color_gradientn(limits = c(-1.5,1.5), colours  = colorRampPalette(rev(brewer.pal(3,"RdBu")))(10)) +
  theme_bw()+theme(axis.text.y = element_text( size = 20),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25)) +
  scale_size(range = c(5, 10))

pdf("Plots/DEG/dot/FCG.pdf", width  = 12, height = 25)
print(plot)
dev.off()

#Full
plot_df <- DEG_df_annotated %>% 
  dplyr::filter(model == "Full" & cell == "Neuron" & comparison %in% c("Ovary_vs_Testis", "Xdose2_vs_Xdose1", "Ydose1_vs_Ydose0", "Ydose2-Ydose0", "Ydose2_vs_Ydose1", "GonadFemale:Xdose2", "GonadFemale:Ydose1", "GonadFemale:Ydose2")) %>%
  dplyr::filter(gene %in% genes_to_plot) %>%
  dplyr::arrange(match(gene, genes_to_plot)) %>%
  mutate(gene = factor(gene, levels = rev(genes_to_plot)))

plot_df$comparison <- factor(plot_df$comparison, levels = c("Ovary_vs_Testis", "Xdose2_vs_Xdose1", "Ydose1_vs_Ydose0", "Ydose2-Ydose0", "Ydose2_vs_Ydose1", "GonadFemale:Xdose2", "GonadFemale:Ydose1", "GonadFemale:Ydose2"))

plot <- ggplot(plot_df, aes(y = gene, x = comparison)) +
  geom_tile(fill = "white") +  xlab("")+ylab("") +
  geom_point(aes(colour = logFC, size = adj.P.Val))  +  
  scale_color_gradientn(limits = c(-1.5,1.5), colours  = colorRampPalette(rev(brewer.pal(3,"RdBu")))(10)) +
  theme_bw()+theme(axis.text.y = element_text( size = 20),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25)) +
  scale_size(range = c(5, 10))

pdf("Plots/DEG/dot/Full.pdf", width  = 12, height = 25)
print(plot)
dev.off()

#Full_continuous
plot_df <- DEG_df_annotated %>% 
  dplyr::filter(model == "Full_continuous" & cell == "Neuron" & comparison %in% c("Ovary_vs_Testis", "Xdose", "Ydose", "GonadFemale:Xdose", "GonadFemale:Ydose", "Xdose:Ydose")) %>%
  dplyr::filter(gene %in% genes_to_plot) %>%
  dplyr::arrange(match(gene, genes_to_plot)) %>%
  mutate(gene = factor(gene, levels = rev(genes_to_plot)))

plot_df$comparison <- factor(plot_df$comparison, levels = c("Ovary_vs_Testis", "Xdose", "Ydose", "GonadFemale:Xdose", "GonadFemale:Ydose", "Xdose:Ydose"))

plot <- ggplot(plot_df, aes(y = gene, x = comparison)) +
  geom_tile(fill = "white") +  xlab("")+ylab("") +
  geom_point(aes(colour = logFC, size = adj.P.Val))  +  
  scale_color_gradientn(limits = c(-1.5,1.5), colours  = colorRampPalette(rev(brewer.pal(3,"RdBu")))(10)) +
  theme_bw()+theme(axis.text.y = element_text( size = 20),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25)) +
  scale_size(range = c(5, 10))

pdf("Plots/DEG/dot/Full_cont.pdf", width  = 12, height = 25)
print(plot)
dev.off()

#XY
plot_df <- DEG_df_annotated %>% 
  mutate(comparison = ifelse(model == c("XY_males"), paste(DEG_df_annotated$comparison, "male", sep = "_"), 
                             ifelse(model == c("XY_females"), paste(DEG_df_annotated$comparison, "female", sep = "_"), 
                                    ifelse(model == c("XY_combined"), paste(DEG_df_annotated$comparison, "combined", sep = "_"),
                                           DEG_df_annotated$comparison))))
   
plot_df <- plot_df %>% 
  dplyr::filter(model %in% c("XY_males", "XY_females", "XY_combined") & cell == "Neuron") %>%
  dplyr::filter(gene %in% genes_to_plot) %>%
  dplyr::arrange(match(gene, genes_to_plot)) %>%
  mutate(gene = factor(gene, levels = rev(genes_to_plot)))

plot_df$comparison <- factor(plot_df$comparison, levels = c("XXY_vs_XY_male", "XXY_vs_XY_female", "XXY_vs_XY_combined", "XYY_vs_XY_male", "XYY_vs_XY_female", "XYY_vs_XY_combined"))

plot <- ggplot(plot_df, aes(y = gene, x = comparison)) +
  geom_tile(fill = "white") +  xlab("")+ylab("") +
  geom_point(aes(colour = logFC, size = adj.P.Val))  +  
  scale_color_gradientn(limits = c(-1.5,1.5), colours  = colorRampPalette(rev(brewer.pal(3,"RdBu")))(10)) +
  theme_bw()+theme(axis.text.y = element_text( size = 20),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25)) +
  scale_size(range = c(5, 10))

pdf("Plots/DEG/dot/XY.pdf", width  = 12, height = 25)
print(plot)
dev.off()

#additionalX
plot_df <- DEG_df_annotated %>% 
  dplyr::filter(model %in% c("additionalX", "additionalX_continuous") & cell == "Neuron" & comparison %in% c("addX1", "addX")) %>%
  dplyr::filter(gene %in% genes_to_plot) %>%
  dplyr::arrange(match(gene, genes_to_plot)) %>%
  mutate(gene = factor(gene, levels = rev(genes_to_plot)))

plot_df$comparison <- factor(plot_df$comparison, levels = c("addX1", "addX"))

plot <- ggplot(plot_df, aes(y = gene, x = comparison)) +
  geom_tile(fill = "white") +  xlab("")+ylab("") +
  geom_point(aes(colour = logFC, size = adj.P.Val))  +  
  scale_color_gradientn(limits = c(-1.5,1.5), colours  = colorRampPalette(rev(brewer.pal(3,"RdBu")))(10)) +
  theme_bw()+theme(axis.text.y = element_text( size = 20),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25)) +
  scale_size(range = c(5, 10)) + 
  scale_y_discrete(drop = FALSE)

pdf("Plots/DEG/dot/addX.pdf", width  = 12, height = 25)
print(plot)
dev.off()

#additionalY
plot_df <- DEG_df_annotated %>% 
  dplyr::filter(model %in% c("additionalY", "additionalY_continuous") & cell == "Neuron" & comparison %in% c("addY1", "addY2", "addY")) %>%
  dplyr::filter(gene %in% genes_to_plot) %>%
  dplyr::arrange(match(gene, genes_to_plot)) %>%
  mutate(gene = factor(gene, levels = rev(genes_to_plot)))

plot_df$comparison <- factor(plot_df$comparison, levels = c("addY1", "addY2", "addY"))

plot <- ggplot(plot_df, aes(y = gene, x = comparison)) +
  geom_tile(fill = "white") +  xlab("")+ylab("") +
  geom_point(aes(colour = logFC, size = adj.P.Val))  +  
  scale_color_gradientn(limits = c(-1.5,1.5), colours  = colorRampPalette(rev(brewer.pal(3,"RdBu")))(10)) +
  theme_bw()+theme(axis.text.y = element_text( size = 20),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25)) +
  scale_size(range = c(5, 10)) + 
  scale_y_discrete(drop = FALSE)

pdf("Plots/DEG/dot/addY.pdf", width  = 12, height = 25)
print(plot)
dev.off()
```

```{r}
##Volcano plots (by cell type)
library(EnhancedVolcano)

allcells <- unique(seurat$cell.type)
allcells <- allcells[!allcells == "Unknown"]

##Normative
for (i in 1:length(allcells)) {
  currentcell <- as.character(allcells[i])
  volcano_list <- list()
  
  plot_df <- DEG_df %>%
    dplyr::filter(model == "Normative" & cell == currentcell)  %>%
    dplyr::mutate(comparison = factor(comparison, levels = c("XXF_vs_XYM")))

  for (j in 1:nlevels(plot_df$comparison)) {
    subset <- plot_df %>% dplyr::filter(comparison == levels(plot_df$comparison)[j])
    volcano_list[[j]] <- EnhancedVolcano(subset,
                            lab = subset$gene,
                            x = 'logFC',
                            y = 'adj.P.Val',
                            subtitle = subset$comparison,
                            FCcutoff = 0.5,
                            titleLabSize = 14,
                            title = currentcell
                            )
  }
  
  pdf(paste0("Plots/DEG/volcano_celltype/Normative_", currentcell, ".pdf"), width  = 15, height = 6)
  print(ggarrange(plotlist = volcano_list, ncol = 3))
  dev.off()
  
}

##FCG
for (i in 1:length(allcells)) {
  currentcell <- as.character(allcells[i])
  volcano_list <- list()
  
  plot_df <- DEG_df %>%
    dplyr::filter(model == "FCG" & cell == currentcell)  %>%
    dplyr::mutate(comparison = factor(comparison, levels = c("Ovary_vs_Testis", "XX_vs_XY", "GonadFemale:SexChromXX")))

  for (j in 1:nlevels(plot_df$comparison)) {
    subset <- plot_df %>% dplyr::filter(comparison == levels(plot_df$comparison)[j])
    volcano_list[[j]] <- EnhancedVolcano(subset,
                            lab = subset$gene,
                            x = 'logFC',
                            y = 'adj.P.Val',
                            subtitle = subset$comparison,
                            FCcutoff = 0.5,
                            titleLabSize = 14,
                            title = currentcell
                            )
  }
  
  pdf(paste0("Plots/DEG/volcano_celltype/FCG_", currentcell, ".pdf"), width  = 15, height = 6)
  print(ggarrange(plotlist = volcano_list, ncol = 3))
  dev.off()
  
}

##Full 
for (i in 1:length(allcells)) {
  currentcell <- as.character(allcells[i])
  volcano_list <- list()
  
  plot_df <- DEG_df %>%
    dplyr::filter(model == "Full" & cell == currentcell)  %>%
    dplyr::mutate(comparison = factor(comparison, levels = c("Ovary_vs_Testis", "Xdose2_vs_Xdose1", "Ydose1_vs_Ydose0", "Ydose2_vs_Ydose0", "Ydose2_vs_Ydose1", "GonadFemale:Xdose2", "GonadFemale:Ydose1", "GonadFemale:Ydose2")))

  for (j in 1:nlevels(plot_df$comparison)) {
    subset <- plot_df %>% dplyr::filter(comparison == levels(plot_df$comparison)[j])
    volcano_list[[j]] <- EnhancedVolcano(subset,
                            lab = subset$gene,
                            x = 'logFC',
                            y = 'adj.P.Val',
                            subtitle = subset$comparison,
                            FCcutoff = 0.5,
                            titleLabSize = 14,
                            title = currentcell
                            )
  }
  
  pdf(paste0("Plots/DEG/volcano_celltype/Full_", currentcell, ".pdf"), width  = 15, height = 18)
  print(ggarrange(plotlist = volcano_list, ncol = 3, nrow = 3))
  dev.off()
  
}

##Full aneuploidy
for (i in 1:length(allcells)) {
  currentcell <- as.character(allcells[i])
  volcano_list <- list()
  
  plot_df <- DEG_df %>%
    dplyr::filter(model == "Full_aneuploidy" & cell == currentcell)  %>%
    dplyr::mutate(comparison = factor(comparison, levels = c("Ovary_vs_Testis", "XX_vs_XY", "Aneuploidy", "GonadFemale:Aneuploidy", "GonadFemale:base_SexChromXX", "base_SexChromXX:Aneuploidy")))

  for (j in 1:nlevels(plot_df$comparison)) {
    subset <- plot_df %>% dplyr::filter(comparison == levels(plot_df$comparison)[j])
    volcano_list[[j]] <- EnhancedVolcano(subset,
                            lab = subset$gene,
                            x = 'logFC',
                            y = 'adj.P.Val',
                            subtitle = subset$comparison,
                            FCcutoff = 0.5,
                            titleLabSize = 14,
                            title = currentcell
                            )
  }
  
  pdf(paste0("Plots/DEG/volcano_celltype/Full_add_", currentcell, ".pdf"), width  = 15, height = 12)
  print(ggarrange(plotlist = volcano_list, ncol = 3, nrow = 2))
  dev.off()
  
}

##Full cont.
for (i in 1:length(allcells)) {
  currentcell <- as.character(allcells[i])
  volcano_list <- list()
  
  plot_df <- DEG_df %>%
    dplyr::filter(model == "Full_continuous" & cell == currentcell)  %>%
    dplyr::mutate(comparison = factor(comparison, levels = c("Ovary_vs_Testis", "Xdose", "Ydose", "GonadFemale:Xdose", "GonadFemale:Ydose", "Xdose:Ydose")))

  for (j in 1:nlevels(plot_df$comparison)) {
    subset <- plot_df %>% dplyr::filter(comparison == levels(plot_df$comparison)[j])
    volcano_list[[j]] <- EnhancedVolcano(subset,
                            lab = subset$gene,
                            x = 'logFC',
                            y = 'adj.P.Val',
                            subtitle = subset$comparison,
                            FCcutoff = 0.5,
                            titleLabSize = 14,
                            title = currentcell
                            )
  }
  
  pdf(paste0("Plots/DEG/volcano_celltype/Full_cont_", currentcell, ".pdf"), width  = 15, height = 12)
  print(ggarrange(plotlist = volcano_list, ncol = 3, nrow = 2))
  dev.off()
  
}

##Full quadratic
for (i in 1:length(allcells)) {
  currentcell <- as.character(allcells[i])
  volcano_list <- list()
  
  plot_df <- DEG_df %>%
    dplyr::filter(model == "Full_quadratic" & cell == currentcell)  %>%
    dplyr::mutate(comparison = factor(comparison, levels = c("Ovary_vs_Testis", "Xdose", "Ydose", "Ydose^2")))

  for (j in 1:nlevels(plot_df$comparison)) {
    subset <- plot_df %>% dplyr::filter(comparison == levels(plot_df$comparison)[j])
    volcano_list[[j]] <- EnhancedVolcano(subset,
                            lab = subset$gene,
                            x = 'logFC',
                            y = 'adj.P.Val',
                            subtitle = subset$comparison,
                            FCcutoff = 0.5,
                            titleLabSize = 14,
                            title = currentcell
                            )
  }
  
  pdf(paste0("Plots/DEG/volcano_celltype/Full_quad_", currentcell, ".pdf"), width  = 20, height = 6)
  print(ggarrange(plotlist = volcano_list, ncol = 4, nrow = 1))
  dev.off()
}

```


##Extra

```{r}
##density plots
library(ggridges)

##FCG
test_df <- DEG_df_annotated %>% dplyr::filter(model %in% c("Normative", "FCG", "Full_continuous"))
test_df2 <- DEG_df_annotated %>% dplyr::filter(model %in% c("Normative", "FCG", "Full_continuous") & adj.P.Val < 0.05)

test_df$log10P <- -log10(test_df$adj.P.Val)
test_df$FC <- 2^test_df$logFC

ggplot(test_df, aes(x = P.Value, y = comparison, fill = comparison)) +
  geom_density_ridges() +
  facet_wrap(~model, ncol = 1) + 
  scale_x_continuous(limits = c(0, 1))

ggplot(test_df, aes(x = adj.P.Val, y = comparison, fill = comparison)) +
  geom_density_ridges() +
  facet_wrap(~model, ncol = 1) + 
  scale_x_continuous(limits = c(0, 1))

ggplot(test_df, aes(x = logFC, y = comparison, fill = comparison)) +
  geom_density_ridges() +
  facet_wrap(~model, ncol = 1) + 
  scale_x_continuous(limits = c(-1.5, 1.5))

ggplot(test_df2, aes(x = logFC, y = comparison, fill = comparison)) +
  geom_density_ridges() +
  facet_wrap(~model, ncol = 1) + 
  scale_x_continuous(limits = c(-1.5, 1.5))
```

```{r}
##cell type proportion tests

library("scProportionTest")
prop_test <- sc_utils(seurat)

prop_test <- permutation_test(
	prop_test, cluster_identity = "cell.type",
	sample_1 = "XYYM", sample_2 = "XYM",
	sample_identity = "Genotype"
)
plot <- permutation_plot(prop_test)

pdf("Plots/DEG/cell_proportion/XYYM_XYM.pdf",width  = 12, height = 5)
print(plot)
dev.off()

prop_test <- permutation_test(
	prop_test, cluster_identity = "cell.type",
	sample_1 = "XYYF", sample_2 = "XYF",
	sample_identity = "Genotype"
)

plot <- permutation_plot(prop_test)

pdf("Plots/DEG/cell_proportion/XYYF_XYF.pdf",width  = 12, height = 5)
print(plot)
dev.off()

prop_test <- permutation_test(
	prop_test, cluster_identity = "cell.type",
	sample_1 = "XXYM", sample_2 = "XXM",
	sample_identity = "Genotype"
)

plot <- permutation_plot(prop_test)

pdf("Plots/DEG/cell_proportion/XXYM_XXM.pdf",width  = 12, height = 5)
print(plot)
dev.off()

prop_test <- permutation_test(
	prop_test, cluster_identity = "cell.type",
	sample_1 = "XXYF", sample_2 = "XXF",
	sample_identity = "Genotype"
)

plot <- permutation_plot(prop_test)

pdf("Plots/DEG/cell_proportion/XYYF_XXF.pdf",width  = 12, height = 5)
print(plot)
dev.off()

prop_test <- permutation_test(
	prop_test, cluster_identity = "cell.type",
	sample_1 = "XXYM", sample_2 = "XYM",
	sample_identity = "Genotype"
)

plot <- permutation_plot(prop_test)

pdf("Plots/DEG/cell_proportion/XXYM_XYM.pdf",width  = 12, height = 5)
print(plot)
dev.off()

prop_test <- permutation_test(
	prop_test, cluster_identity = "cell.type",
	sample_1 = "XXYF", sample_2 = "XYF",
	sample_identity = "Genotype"
)

plot <- permutation_plot(prop_test)

pdf("Plots/DEG/cell_proportion/XXYF_XYF.pdf",width  = 12, height = 5)
print(plot)
dev.off()
```



