---
title: "DEG3"
output: html_document
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/u/project/xyang123/julianas/medial_septum_sct")
```

```{r}
library(limma)
library(edgeR)
library(data.table)
library(plyr)
library(dplyr)
library(fastDummies)
library(Seurat)
library(clusterProfiler)
library(enrichplot)
library(org.Mm.eg.db)
library(ggplot2)
library(RColorBrewer)
library(grid)
library(gridExtra)
library(SingleCellExperiment)
#library(Matrix.utils)
library(magrittr)
library(simplifyEnrichment)
library(stringr)
library(enrichplot)
library(EnhancedVolcano)
library(biomaRt)
library(readxl)
library(UpSetR)
library(tidyr)
```

```{r}
seurat <- readRDS( "Saves/For_Figures.rds") %>% 
    DietSeurat(counts = TRUE)
```

```{r}
gene_annotation <- read.table("Gene_annotation/gene_annotation_final.txt", sep = "\t")
#gene_annotation <- read.table("Gene_annotation/gene_annotation_biomart.txt", sep = "\t")

#xescape_brain <- read_xlsx("Gene_annotation/xescape_brain.xlsx")
#colnames(xescape_brain) <- xescape_brain[1, ]
#xescape_brain <- xescape_brain[-1, ]

#gene_annotation$gene_class <- ifelse(gene_annotation$Chr %in% c("X", "Y"), "Sex-linked", "Autosome")
#gene_annotation$Xescape <- ifelse(gene_annotation$Input %in% xescape_brain$Gene, "Escapee", "Non-escapee")

par_genes <- c("Mid1", "Sts", "Nlgn4", "Asmt")
```

```{r}
#clean up results - make loop that cycles over all cell types? 
allcells <- unique(seurat$cell.type)
allcells <- allcells[!allcells == "Unknown"]

alldata <- list.dirs("Data/Derived/DEG", recursive = TRUE)
alldata <- alldata[-1]
currentcell <- allcells[1]
for (i in 1:length(alldata)) {
  load(paste0(alldata[i], "/", currentcell, ".rda"))
}

normative_df <- bind_rows(resultlist_normative, .id = "comparison") %>%
  dplyr::mutate(model = "normative",
                celltype = currentcell) %>%
  tibble::rownames_to_column(var = "Gene")

fcg_df <- bind_rows(resultlist_fcg, .id = "comparison") %>%
  dplyr::mutate(model = "FCG",
                celltype = currentcell) %>%
  tibble::rownames_to_column(var = "Gene")
fcg_df$Gene <- gsub("\\...*", "", fcg_df$Gene)

dose_df <- bind_rows(resultlist_dosage, .id = "comparison") %>%
  dplyr::mutate(model = "Dose",
                celltype = currentcell) %>%
  tibble::rownames_to_column(var = "Gene")
dose_df$Gene <- gsub("\\...*", "", dose_df$Gene)

XY_male_df <- bind_rows(resultlist_XY, .id = "comparison") %>%
  dplyr::mutate(model = "XY_male",
                celltype = currentcell) %>%
  tibble::rownames_to_column(var = "Gene")
XY_male_df$Gene <- gsub("\\...*", "", XY_male_df$Gene)

interaction_genes <- bind_rows(resultlist_XY_comb[3:6], .id = "comparison") %>%
  dplyr::filter(adj.P.Val < 0.05) %>%
  tibble::rownames_to_column(var = "Gene") %>%
  dplyr::select(Gene) %>%
  unlist(use.names = FALSE)
interaction_genes <- gsub("\\...*", "", interaction_genes)
interaction_genes <- interaction_genes[!duplicated(interaction_genes)]

XY_comb_df <- bind_rows(resultlist_XY_comb[1:2], .id = "comparison") %>%
  dplyr::mutate(model = "XY_comb",
                celltype = currentcell) %>%
  tibble::rownames_to_column(var = "Gene")
XY_comb_df$Gene <- gsub("\\...*", "", XY_comb_df$Gene)
XY_comb_df <- XY_comb_df %>% dplyr::filter(!Gene %in% interaction_genes)

addX_df <- bind_rows(resultlist_additionalX, .id = "comparison") %>%
  dplyr::mutate(model = "additionalX",
                celltype = currentcell) %>%
  tibble::rownames_to_column(var = "Gene")
addX_df$Gene <- gsub("\\...*", "", addX_df$Gene)

addY_df <- bind_rows(resultlist_additionalY, .id = "comparison") %>%
  dplyr::mutate(model = "additionalY",
                celltype = currentcell) %>%
  tibble::rownames_to_column(var = "Gene")
addY_df$Gene <- gsub("\\...*", "", addY_df$Gene)

combined_df <- rbind(normative_df, fcg_df, dose_df, XY_male_df, XY_comb_df, addX_df, addY_df)

combined_df <- combined_df %>% dplyr::filter(adj.P.Val < 0.05 & abs(logFC) > 0.5) %>% 
  left_join(gene_annotation, by = c("Gene"="Input"), multiple = "all")
combined_df$Chr <- ifelse(combined_df$Gene == "AC149090", "unplaced scaffold", combined_df$Chr)
```

```{r}
allcells <- unique(seurat$cell.type)
allcells <- allcells[!allcells == "Unknown"]

alldata <- list.dirs("Data/Derived/DEG", recursive = TRUE)
alldata <- alldata[-1]

suppressWarnings(rm(combined_df_all))

for(i in 1:length(allcells)) {
  currentcell <- allcells[i]
  
  for(i in 1:length(alldata)) {
    load(paste0(alldata[i], "/", currentcell, ".rda"))
  }
  
  normative_df <- bind_rows(resultlist_normative, .id = "comparison") %>% 
    dplyr::mutate(model = "normative", 
                  celltype = currentcell) %>% 
    tibble::rownames_to_column(var = "Gene")

  fcg_df <- bind_rows(resultlist_fcg, .id = "comparison") %>%
    dplyr::mutate(model = "FCG",
                  celltype = currentcell) %>%
    tibble::rownames_to_column(var = "Gene")
  fcg_df$Gene <- gsub("\\...*", "", fcg_df$Gene)

  dose_df <- bind_rows(resultlist_dosage, .id = "comparison") %>%
    dplyr::mutate(model = "Dose",
                  celltype = currentcell) %>%
    tibble::rownames_to_column(var = "Gene")
  dose_df$Gene <- gsub("\\...*", "", dose_df$Gene)
  
  XY_male_df <- bind_rows(resultlist_XY, .id = "comparison") %>%
    dplyr::mutate(model = "XY_male",
                  celltype = currentcell) %>%
    tibble::rownames_to_column(var = "Gene")
  XY_male_df$Gene <- gsub("\\...*", "", XY_male_df$Gene)
  
  interaction_genes <- bind_rows(resultlist_XY_comb[3:6], .id = "comparison") %>%
    dplyr::filter(adj.P.Val < 0.05) %>%
    tibble::rownames_to_column(var = "Gene") %>%
    dplyr::select(Gene) %>%
    unlist(use.names = FALSE)
  interaction_genes <- gsub("\\...*", "", interaction_genes)
  interaction_genes <- interaction_genes[!duplicated(interaction_genes)]
  
  XY_comb_df <- bind_rows(resultlist_XY_comb[1:2], .id = "comparison") %>%
    dplyr::mutate(model = "XY_comb",
                  celltype = currentcell) %>%
    tibble::rownames_to_column(var = "Gene")
  XY_comb_df$Gene <- gsub("\\...*", "", XY_comb_df$Gene)
  XY_comb_df <- XY_comb_df %>% dplyr::filter(!Gene %in% interaction_genes)
  
  addX_df <- bind_rows(resultlist_additionalX, .id = "comparison") %>%
    dplyr::mutate(model = "additionalX",
                  celltype = currentcell) %>%
    tibble::rownames_to_column(var = "Gene")
  addX_df$Gene <- gsub("\\...*", "", addX_df$Gene)
  
  addY_df <- bind_rows(resultlist_additionalY, .id = "comparison") %>%
    dplyr::mutate(model = "additionalY",
                  celltype = currentcell) %>%
    tibble::rownames_to_column(var = "Gene")
  addY_df$Gene <- gsub("\\...*", "", addY_df$Gene)
  
  combined_df <- rbind(normative_df, fcg_df, dose_df, XY_male_df, XY_comb_df, addX_df, addY_df)
  
  combined_df <- combined_df %>% dplyr::filter(adj.P.Val < 0.05 & abs(logFC) > 0.1) %>% 
    left_join(gene_annotation, by = c("Gene"="Input"), multiple = "all")
  
  if(!exists("combined_df_all")){
      combined_df_all <- combined_df
    }else{
      combined_df_all <- rbind.data.frame(combined_df_all, combined_df) }

}

table(combined_df_all$model)
```

```{r}
allcells <- unique(seurat$cell.type)
allcells <- allcells[!allcells == "Unknown"]

alldata <- list.dirs("Data/Derived/DEG", recursive = TRUE)
alldata <- alldata[-1]

topgenes_up <- list()
topgenes_down <- list()
DEG_up_results <- list()
DEG_down_results <- list()
pathway_up_results <- list()
pathway_down_results <- list()

organism <- org.Mm.eg.db

gene_list <- rownames(seurat@assays$RNA@counts)
gene_list <- na.omit(gene_list)

##pathway analysis - OVERREPRESENTATION ANALYSIS 
#Upregulated DEGs
for(i in 2:length(allcells)) {
  currentcell <- as.character(allcells[i])
  
  for(j in 1:length(alldata)) {
    load(paste0(alldata[j], "/", currentcell, ".rda"))
  }
  
  ##normative model
  for(k in 1:length(resultlist_normative)) {

    currentframe <- resultlist_normative[[k]]
    currentcondition <- names(resultlist_normative[k])
    currentmodel <- "Normative"
    currentsign <- "Upregulated"
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    currentframe <- currentframe %>% dplyr::filter(logFC > 0 & adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentframe$sign <- currentsign

    topgenes_up[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    
    genes <- currentframe %>% dplyr::select(gene) %>% unlist()
    go_enrich <- enrichGO(gene = genes,
                      universe = gene_list,
                      OrgDb = organism, 
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)
    
    if(is.null(go_enrich)) {next}
    
    fold <- NULL
    for(r in 1:nrow(go_enrich@result)){
      enriched_genes <- unlist(strsplit(go_enrich@result$geneID[r],"/"))
      enriched_genes <- enriched_genes[enriched_genes %in% currentframe$gene]
      fold[r] <- median(currentframe[currentframe$gene %in% enriched_genes, "logFC"], na.rm = T) }
    
    currentpath <- go_enrich@result
    currentpath$fold <- fold 
    currentpath$sign <- currentsign

    DEG_up_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe
    pathway_up_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentpath
  }
  
  ##FCG model 
  for(k in 1:length(resultlist_fcg)) {
 
    currentframe <- resultlist_fcg[[k]]
    currentcondition <- names(resultlist_fcg[k])
    currentmodel <- "FCG"
    currentsign <- "Upregulated"
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    currentframe <- currentframe %>% dplyr::filter(logFC > 0 & adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentframe$sign <- currentsign

    topgenes_up[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    
    genes <- currentframe %>% dplyr::select(gene) %>% unlist()
    go_enrich <- enrichGO(gene = genes,
                      universe = gene_list,
                      OrgDb = organism, 
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)
    
    if(is.null(go_enrich)) {next}
    
    fold <- NULL
    for(r in 1:nrow(go_enrich@result)){
      enriched_genes <- unlist(strsplit(go_enrich@result$geneID[r],"/"))
      enriched_genes <- enriched_genes[enriched_genes %in% currentframe$gene]
      fold[r] <- median(currentframe[currentframe$gene %in% enriched_genes, "logFC"], na.rm = T) }
    
    currentpath <- go_enrich@result
    currentpath$fold <- fold 
    currentpath$sign <- currentsign

    DEG_up_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe
    pathway_up_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentpath
  }
  
  ##full model 
  for(k in 1:length(resultlist_dosage)) {

    currentframe <- resultlist_dosage[[k]]
    currentcondition <- names(resultlist_dosage[k])
    currentmodel <- "Dose"
    currentsign <- "Upregulated"
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    currentframe <- currentframe %>% dplyr::filter(logFC > 0 & adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentframe$sign <- currentsign

    topgenes_up[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    
    genes <- currentframe %>% dplyr::select(gene) %>% unlist()
    go_enrich <- enrichGO(gene = genes,
                      universe = gene_list,
                      OrgDb = organism, 
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)
    
    if(is.null(go_enrich)) {next}
    
    fold <- NULL
    for(r in 1:nrow(go_enrich@result)){
      enriched_genes <- unlist(strsplit(go_enrich@result$geneID[r],"/"))
      enriched_genes <- enriched_genes[enriched_genes %in% currentframe$gene]
      fold[r] <- median(currentframe[currentframe$gene %in% enriched_genes, "logFC"], na.rm = T) }
    
    currentpath <- go_enrich@result
    currentpath$fold <- fold 
    currentpath$sign <- currentsign

    DEG_up_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe
    pathway_up_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentpath
  }
  
  ##XY model 
  for(k in 1:length(resultlist_XY)) {

    currentframe <- resultlist_XY[[k]]
    currentcondition <- names(resultlist_XY[k])
    currentmodel <- "XY_males"
    currentsign <- "Upregulated"
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    currentframe <- currentframe %>% dplyr::filter(logFC > 0 & adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentframe$sign <- currentsign

    topgenes_up[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    
    genes <- currentframe %>% dplyr::select(gene) %>% unlist()
    go_enrich <- enrichGO(gene = genes,
                      universe = gene_list,
                      OrgDb = organism, 
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)
    
    if(is.null(go_enrich)) {next}
    
    fold <- NULL
    for(r in 1:nrow(go_enrich@result)){
      enriched_genes <- unlist(strsplit(go_enrich@result$geneID[r],"/"))
      enriched_genes <- enriched_genes[enriched_genes %in% currentframe$gene]
      fold[r] <- median(currentframe[currentframe$gene %in% enriched_genes, "logFC"], na.rm = T) }
    
    currentpath <- go_enrich@result
    currentpath$fold <- fold 
    currentpath$sign <- currentsign

    DEG_up_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe
    pathway_up_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentpath
  }
  
  ##XY model (combined) 
  for(k in 1:2) {
    
    currentframe <- resultlist_XY_comb[[k]]
    currentcondition <- names(resultlist_XY_comb[k])
    currentmodel <- "XY_comb"
    currentsign <- "Upregulated"
    if(nrow(currentframe) == 0) {next}
    
    interaction_genes <- bind_rows(resultlist_XY_comb[3:6], .id = "comparison") %>% 
      dplyr::filter(adj.P.Val < 0.05) %>%
      tibble::rownames_to_column(var = "gene") %>%
      dplyr::select(gene) %>%
      unlist(use.names = FALSE)
    interaction_genes <- gsub("\\...*", "", interaction_genes)
    interaction_genes <- interaction_genes[!duplicated(interaction_genes)]
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    currentframe <- currentframe %>% dplyr::filter(logFC > 0 & adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!gene %in% c(interaction_genes, "Gm42418"))
    if(nrow(currentframe) == 0) {next}
    currentframe$sign <- currentsign

    topgenes_up[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    
    genes <- currentframe %>% dplyr::select(gene) %>% unlist()
    go_enrich <- enrichGO(gene = genes,
                      universe = gene_list,
                      OrgDb = organism, 
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)
    
    if(is.null(go_enrich)) {next}
    
    fold <- NULL
    for(r in 1:nrow(go_enrich@result)){
      enriched_genes <- unlist(strsplit(go_enrich@result$geneID[r],"/"))
      enriched_genes <- enriched_genes[enriched_genes %in% currentframe$gene]
      fold[r] <- median(currentframe[currentframe$gene %in% enriched_genes, "logFC"], na.rm = T) }
    
    currentpath <- go_enrich@result
    currentpath$fold <- fold 
    currentpath$sign <- currentsign

    DEG_up_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe
    pathway_up_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentpath
  }
    
  
  ##AdditionalX model 
  for(k in 1:length(resultlist_additionalX)) {

    currentframe <- resultlist_additionalX[[k]]
    currentcondition <- names(resultlist_additionalX[k])
    currentmodel <- "additionalX"
    currentsign <- "Upregulated"
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    currentframe <- currentframe %>% dplyr::filter(logFC > 0 & adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentframe$sign <- currentsign

    topgenes_up[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    
    genes <- currentframe %>% dplyr::select(gene) %>% unlist()
    go_enrich <- enrichGO(gene = genes,
                      universe = gene_list,
                      OrgDb = organism, 
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)
    
    if(is.null(go_enrich)) {next}
    
    fold <- NULL
    for(r in 1:nrow(go_enrich@result)){
      enriched_genes <- unlist(strsplit(go_enrich@result$geneID[r],"/"))
      enriched_genes <- enriched_genes[enriched_genes %in% currentframe$gene]
      fold[r] <- median(currentframe[currentframe$gene %in% enriched_genes, "logFC"], na.rm = T) }
    
    currentpath <- go_enrich@result
    currentpath$fold <- fold 
    currentpath$sign <- currentsign

    DEG_up_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe
    pathway_up_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentpath
  }
  
  ##AdditionalY model 
  for(k in 1:length(resultlist_additionalY)) {

    currentframe <- resultlist_additionalY[[k]]
    currentcondition <- names(resultlist_additionalY[k])
    currentmodel <- "additionalY"
    currentsign <- "Upregulated"
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    currentframe <- currentframe %>% dplyr::filter(logFC > 0 & adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentframe$sign <- currentsign

    topgenes_up[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    
    genes <- currentframe %>% dplyr::select(gene) %>% unlist()
    go_enrich <- enrichGO(gene = genes,
                      universe = gene_list,
                      OrgDb = organism, 
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)
    
    if(is.null(go_enrich)) {next}
    
    fold <- NULL
    for(r in 1:nrow(go_enrich@result)){
      enriched_genes <- unlist(strsplit(go_enrich@result$geneID[r],"/"))
      enriched_genes <- enriched_genes[enriched_genes %in% currentframe$gene]
      fold[r] <- median(currentframe[currentframe$gene %in% enriched_genes, "logFC"], na.rm = T) }
    
    currentpath <- go_enrich@result
    currentpath$fold <- fold 
    currentpath$sign <- currentsign

    DEG_up_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe
    pathway_up_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentpath
  }
}

save(DEG_up_results, pathway_up_results, topgenes_up, file = "Saves/DEG_pathway_up.Rda")
```

```{r}
##pathway analysis - OVERREPRESENTATION ANALYSIS 
#Downregulated DEGs
for(i in 1:length(allcells)) {
  currentcell <- as.character(allcells[i])
  
  for(j in 1:length(alldata)) {
    load(paste0(alldata[j], "/", currentcell, ".rda"))
  }
  
  ##normative model
  for(k in 1:length(resultlist_normative)) {

    currentframe <- resultlist_normative[[k]]
    currentcondition <- names(resultlist_normative[k])
    currentmodel <- "Normative"
    currentsign <- "Downregulated"
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    currentframe <- currentframe %>% dplyr::filter(logFC < 0 & adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentframe$sign <- currentsign

    topgenes_down[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    
    genes <- currentframe %>% dplyr::select(gene) %>% unlist()
    go_enrich <- enrichGO(gene = genes,
                      universe = gene_list,
                      OrgDb = organism, 
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)
    
    if(is.null(go_enrich)) {next}
    
    fold <- NULL
    for(r in 1:nrow(go_enrich@result)){
      enriched_genes <- unlist(strsplit(go_enrich@result$geneID[r],"/"))
      enriched_genes <- enriched_genes[enriched_genes %in% currentframe$gene]
      fold[r] <- median(currentframe[currentframe$gene %in% enriched_genes, "logFC"], na.rm = T) }
    
    currentpath <- go_enrich@result
    currentpath$fold <- fold 
    currentpath$sign <- currentsign

    DEG_down_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe
    pathway_down_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentpath
  }
  
  ##FCG model 
  for(k in 1:length(resultlist_fcg)) {
 
    currentframe <- resultlist_fcg[[k]]
    currentcondition <- names(resultlist_fcg[k])
    currentmodel <- "FCG"
    currentsign <- "Downregulated"
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    currentframe <- currentframe %>% dplyr::filter(logFC < 0 & adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentframe$sign <- currentsign

    topgenes_down[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    
    genes <- currentframe %>% dplyr::select(gene) %>% unlist()
    go_enrich <- enrichGO(gene = genes,
                      universe = gene_list,
                      OrgDb = organism, 
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)
    
    if(is.null(go_enrich)) {next}
    
    fold <- NULL
    for(r in 1:nrow(go_enrich@result)){
      enriched_genes <- unlist(strsplit(go_enrich@result$geneID[r],"/"))
      enriched_genes <- enriched_genes[enriched_genes %in% currentframe$gene]
      fold[r] <- median(currentframe[currentframe$gene %in% enriched_genes, "logFC"], na.rm = T) }
    
    currentpath <- go_enrich@result
    currentpath$fold <- fold 
    currentpath$sign <- currentsign

    DEG_down_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe
    pathway_down_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentpath
  }
  
  ##full model 
  for(k in 1:length(resultlist_dosage)) {

    currentframe <- resultlist_dosage[[k]]
    currentcondition <- names(resultlist_dosage[k])
    currentmodel <- "Dose"
    currentsign <- "Downregulated"
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    currentframe <- currentframe %>% dplyr::filter(logFC < 0 & adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentframe$sign <- currentsign

    topgenes_down[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    
    genes <- currentframe %>% dplyr::select(gene) %>% unlist()
    go_enrich <- enrichGO(gene = genes,
                      universe = gene_list,
                      OrgDb = organism, 
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)
    
    if(is.null(go_enrich)) {next}
    
    fold <- NULL
    for(r in 1:nrow(go_enrich@result)){
      enriched_genes <- unlist(strsplit(go_enrich@result$geneID[r],"/"))
      enriched_genes <- enriched_genes[enriched_genes %in% currentframe$gene]
      fold[r] <- median(currentframe[currentframe$gene %in% enriched_genes, "logFC"], na.rm = T) }
    
    currentpath <- go_enrich@result
    currentpath$fold <- fold 
    currentpath$sign <- currentsign

    DEG_down_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe
    pathway_down_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentpath
  }
  
  ##XY model 
  for(k in 1:length(resultlist_XY)) {

    currentframe <- resultlist_XY[[k]]
    currentcondition <- names(resultlist_XY[k])
    currentmodel <- "XY_males"
    currentsign <- "Downregulated"
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    currentframe <- currentframe %>% dplyr::filter(logFC < 0 & adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentframe$sign <- currentsign

    topgenes_down[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    
    genes <- currentframe %>% dplyr::select(gene) %>% unlist()
    go_enrich <- enrichGO(gene = genes,
                      universe = gene_list,
                      OrgDb = organism, 
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)
    
    if(is.null(go_enrich)) {next}
    
    fold <- NULL
    for(r in 1:nrow(go_enrich@result)){
      enriched_genes <- unlist(strsplit(go_enrich@result$geneID[r],"/"))
      enriched_genes <- enriched_genes[enriched_genes %in% currentframe$gene]
      fold[r] <- median(currentframe[currentframe$gene %in% enriched_genes, "logFC"], na.rm = T) }
    
    currentpath <- go_enrich@result
    currentpath$fold <- fold 
    currentpath$sign <- currentsign

    DEG_down_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe
    pathway_down_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentpath
  }
  
  ##XY model (combined) 
  for(k in 1:2) {
    
    currentframe <- resultlist_XY_comb[[k]]
    currentcondition <- names(resultlist_XY_comb[k])
    currentmodel <- "XY_comb"
    currentsign <- "Downregulated"
    if(nrow(currentframe) == 0) {next}
    
    interaction_genes <- bind_rows(resultlist_XY_comb[3:6], .id = "comparison") %>% 
      dplyr::filter(adj.P.Val < 0.05) %>%
      tibble::rownames_to_column(var = "gene") %>%
      dplyr::select(gene) %>%
      unlist(use.names = FALSE)
    interaction_genes <- gsub("\\...*", "", interaction_genes)
    interaction_genes <- interaction_genes[!duplicated(interaction_genes)]
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    currentframe <- currentframe %>% dplyr::filter(logFC < 0 & adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!gene %in% c(interaction_genes, "Gm42418"))
    if(nrow(currentframe) == 0) {next}
    currentframe$sign <- currentsign

    topgenes_down[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    
    genes <- currentframe %>% dplyr::select(gene) %>% unlist()
    go_enrich <- enrichGO(gene = genes,
                      universe = gene_list,
                      OrgDb = organism, 
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)
    
    if(is.null(go_enrich)) {next}
    
    fold <- NULL
    for(r in 1:nrow(go_enrich@result)){
      enriched_genes <- unlist(strsplit(go_enrich@result$geneID[r],"/"))
      enriched_genes <- enriched_genes[enriched_genes %in% currentframe$gene]
      fold[r] <- median(currentframe[currentframe$gene %in% enriched_genes, "logFC"], na.rm = T) }
    
    currentpath <- go_enrich@result
    currentpath$fold <- fold 
    currentpath$sign <- currentsign

    DEG_down_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe
    pathway_down_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentpath
  }
    
  
  ##AdditionalX model 
  for(k in 1:length(resultlist_additionalX)) {

    currentframe <- resultlist_additionalX[[k]]
    currentcondition <- names(resultlist_additionalX[k])
    currentmodel <- "additionalX"
    currentsign <- "Downregulated"
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    currentframe <- currentframe %>% dplyr::filter(logFC < 0 & adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentframe$sign <- currentsign

    topgenes_down[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    
    genes <- currentframe %>% dplyr::select(gene) %>% unlist()
    go_enrich <- enrichGO(gene = genes,
                      universe = gene_list,
                      OrgDb = organism, 
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)
    
    if(is.null(go_enrich)) {next}
    
    fold <- NULL
    for(r in 1:nrow(go_enrich@result)){
      enriched_genes <- unlist(strsplit(go_enrich@result$geneID[r],"/"))
      enriched_genes <- enriched_genes[enriched_genes %in% currentframe$gene]
      fold[r] <- median(currentframe[currentframe$gene %in% enriched_genes, "logFC"], na.rm = T) }
    
    currentpath <- go_enrich@result
    currentpath$fold <- fold 
    currentpath$sign <- currentsign

    DEG_down_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe
    pathway_down_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentpath
  }
  
  ##AdditionalY model 
  for(k in 1:length(resultlist_additionalY)) {

    currentframe <- resultlist_additionalY[[k]]
    currentcondition <- names(resultlist_additionalY[k])
    currentmodel <- "additionalY"
    currentsign <- "Downregulated"
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    currentframe <- currentframe %>% dplyr::filter(logFC < 0 & adj.P.Val < 0.05)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentframe$sign <- currentsign

    topgenes_down[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))]
    
    genes <- currentframe %>% dplyr::select(gene) %>% unlist()
    go_enrich <- enrichGO(gene = genes,
                      universe = gene_list,
                      OrgDb = organism, 
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)
    
    if(is.null(go_enrich)) {next}
    
    fold <- NULL
    for(r in 1:nrow(go_enrich@result)){
      enriched_genes <- unlist(strsplit(go_enrich@result$geneID[r],"/"))
      enriched_genes <- enriched_genes[enriched_genes %in% currentframe$gene]
      fold[r] <- median(currentframe[currentframe$gene %in% enriched_genes, "logFC"], na.rm = T) }
    
    currentpath <- go_enrich@result
    currentpath$fold <- fold 
    currentpath$sign <- currentsign

    DEG_down_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentframe
    pathway_down_results[[currentmodel]][[currentcondition]][[currentcell]] <- currentpath
  }
}

save(DEG_down_results, pathway_down_results, topgenes_down, file = "Saves/DEG_pathway_down.Rda")
```


```{r}
test <- unlist(DEG_up_results, use.names = TRUE, recursive = FALSE) %>%
  unlist(use.names=TRUE, recursive = FALSE)
test <- map2(test, names(test), ~ mutate(.x, model.condition.cell = .y))
final <- bind_rows(test)

str_split(final$model.condition.cell[1], "\\.")[1]

```

```{r}
gene_list <- combined_df_all %>% dplyr::filter(comparison %in% c("XXF_vs_XYM", "SexChromXX", "Xdose2", "Ydose1", "Ydose2", "XXY_vs_XY", "XYY_vs_XY") | (model == "FCG" & comparison %in% c("GonadMale", "GonadFemale")))
gene_list <- as.data.frame(table(gene_list$Gene, gene_list$comparison))
gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq)
gene_list[2:10] <- ifelse(gene_list[2:10] == 0, 0, 1)
gene_list[2:10] <- lapply(gene_list[2:10], factor)

set_vars <- c("XXF_vs_XYM", "GonadMale", "GonadFemale", "SexChromXX", "XXY_vs_XY", "XYY_vs_XY", "Xdose2", "Ydose1", "Ydose2")
plot <- UpSetR::upset(data = gene_list, sets = set_vars)
plot

pdf("Markers/SoupX/pathway/cell_type/upsetplot_0_1_2_model.pdf",width  = 10, height = 5)
print(plot)
dev.off()
```


```{r}
barplotframe <- as.data.frame(table(resultlist$Gene_class))
barplotframe$Var1 <- as.character(barplotframe$Var1)
barplotframe <- barplotframe %>%
  dplyr::mutate(Class = Var1)

ggplot(barplotframe, aes(Class, Freq)) + 
  geom_bar(position = "dodge", stat = "identity")

barplotframe <- as.data.frame(table(resultlist$Xescape))
barplotframe$Var1 <- as.character(barplotframe$Var1)
barplotframe <- barplotframe %>%
  dplyr::mutate(Class = Var1)

ggplot(barplotframe, aes(Class, Freq)) + 
  geom_bar(position = "dodge", stat = "identity")

pdf("Markers/SoupX/pathway/cell_type/additionalY_0_1_2_model/DEG_bar.pdf",width = 15,height = 5)
print()


##why these steps?
combined_all_frame$FDR <- -log10(combined_all_frame$adj.P.Val)
combined_all_frame$FDR[combined_all_frame$FDR > 3] <- 3
combined_all_frame$logFC2 <- combined_all_frame$logFC
combined_all_frame$logFC2[combined_all_frame$logFC2 > 3] <- 3
combined_all_frame$logFC2[combined_all_frame$logFC2 < -3] <- -3

combined_all_pathway_frame$FDR <- -log10(combined_all_pathway_frame$p.adjust)
combined_all_pathway_frame$FDR[combined_all_pathway_frame$FDR > 3] <- 3
combined_all_pathway_frame$fold[combined_all_pathway_frame$fold > 1] <- 1
combined_all_pathway_frame$fold[combined_all_pathway_frame$fold < -1] <- -1

#combined_all_pathway_frame$Term <- gsub(" \\(.*\\)","",combined_all_pathway_frame$Term)
genesofinterest <- c("Xist", "Eif2s3x", "Ddx3x", "Kdm5c", "Kdm6a", "Uty", "Kdm5d", "Ddx3y", "Eif2s3y")
```

```{r}
##visualize DEG results: volcano plots 


##highlight genes that are sex-linked, escapees, and PAR? 

```

```{r}
##pathway analysis final loop - OVERREPRESENTATION ANALYSIS 
#setwd("Markers/SoupX/DEG/no_contrasts/cell_type2")
allfiles <- list.files("Markers/SoupX/DEG/no_contrasts/cell_type2")
topgenes <- list()
enrichlist <- list()
organism <- "org.Mm.eg.db"

suppressWarnings(rm(combined_all_frame, combined_all_pathway_frame))

##Test loop 
for(i in 1:length(allfiles)) {
  currentfile <- allfiles[i]
  currentcell <- gsub(".rda","",unlist(currentfile),fixed = T)
  load(file.path("/u/project/xyang123/julianas/singlecell_cerebellum/Markers/SoupX/DEG/no_contrasts/cell_type2", currentfile))
  
  gene_list <- rownames(seurat@assays$RNA@counts)
  gene_list <- na.omit(gene_list) ##remove lowly expressed genes? 
  
  ##pathway enrichment for additionalX effect 
  for(j in 1:length(resultlist_additionalX)) {
    suppressWarnings(rm(currentframe, currentcondition))
    currentframe <- resultlist_additionalX[[j]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    currentframe <- currentframe %>%
      dplyr::filter(abs(logFC) > 0.1 & logFC > 0) %>%
      dplyr::arrange(adj.P.Val)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcondition <- names(resultlist_additionalX[j])
    currentframe$cell <- currentcell
    currentframe$condition <- currentcondition
    currentframe$sign <- "Upregulated"

    topgenes[[currentcondition]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))] ##take at most 3 
    
    sig_genes_df = subset(currentframe, adj.P.Val < 0.05 & logFC > 0) 
    genes <- sig_genes_df$logFC
    names(genes) <- sig_genes_df$gene
    genes <- na.omit(genes)
    genes <- names(genes)
    ##genes <- names(genes)[abs(genes) > 2]
    
    go_enrich <- enrichGO(gene = genes,
                      universe = gene_list,
                      OrgDb = organism, 
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)
    
    #enrichlist[[currentcondition]][[currentcell]][["UP"]] <- go_enrich_up
    
    if(is.null(go_enrich)) {next}
    
    fold <- NULL
    for(r in 1:nrow(go_enrich@result)){
      enriched_genes <- unlist(strsplit(go_enrich@result$geneID[r],"/"))
      enriched_genes <- enriched_genes[enriched_genes %in% currentframe$gene]
      fold[r] <- median(currentframe[currentframe$gene %in% enriched_genes, "logFC"], na.rm = T) }
    
    currentpath <- go_enrich@result
    currentpath$fold <- fold 
    currentpath$condition <- currentcondition
    currentpath$cell <- currentcell
    currentpath$sign <- "Upregulated"
    
    #save(go_enrich_up, go_enrich_down, file = paste0("/u/home/j/julianas/project-xyang123/singlecell_cerebellum/Markers/SoupX/pathway/saves/", currentcondition, currentcell, ".Rda"))

    if(!exists("combined_all_pathway_frame")) {
      combined_all_pathway_frame <- currentpath 
      } else {
      combined_all_pathway_frame <- rbind.data.frame(combined_all_pathway_frame, currentpath) }
    
    if(!exists("combined_all_frame")){
        combined_all_frame <- currentframe
      }else{
        combined_all_frame <- rbind.data.frame(combined_all_frame, currentframe) }

    
  }
  
  ##pathway enrichment for additionalY effect 
  for(j in 1:length(resultlist_additionalY)) {
    suppressWarnings(rm(currentframe, currentcondition))
    currentframe <- resultlist_additionalY[[j]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    currentframe <- currentframe %>%
      dplyr::filter(abs(logFC) > 0.1 & logFC > 0) %>%
      dplyr::arrange(adj.P.Val)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcondition <- names(resultlist_additionalY[j])
    currentframe$cell <- currentcell
    currentframe$condition <- currentcondition
    currentframe$sign <- "Upregulated"
    
    topgenes[[currentcondition]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))] ##take at most 3 
    
    sig_genes_df = subset(currentframe, adj.P.Val < 0.05 & logFC > 0) 
    genes <- sig_genes_df$logFC
    names(genes) <- sig_genes_df$gene
    genes <- na.omit(genes)
    genes <- names(genes)
    ##genes <- names(genes)[abs(genes) > 2]
    
    go_enrich <- enrichGO(gene = genes,
                      universe = gene_list,
                      OrgDb = organism, 
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)
    
    #enrichlist[[currentcondition]][[currentcell]][["UP"]] <- go_enrich_up
    
    if(is.null(go_enrich)) {next}
    
    fold <- NULL
    for(r in 1:nrow(go_enrich@result)){
      enriched_genes <- unlist(strsplit(go_enrich@result$geneID[r],"/"))
      enriched_genes <- enriched_genes[enriched_genes %in% currentframe$gene]
      fold[r] <- median(currentframe[currentframe$gene %in% enriched_genes, "logFC"], na.rm = T) }
    
    #currentpath <- go_enrich_up@result %>% dplyr::filter(p.adjust < 0.05 & qvalue < 0.05)
    currentpath <- go_enrich@result
    currentpath$fold <- fold 
    currentpath$condition <- currentcondition
    currentpath$cell <- currentcell
    currentpath$sign <- "Upregulated"

    #save(go_enrich_up, go_enrich_down, file = paste0("/u/home/j/julianas/project-xyang123/singlecell_cerebellum/Markers/SoupX/pathway/saves/", currentcondition, currentcell, ".Rda"))

    if(!exists("combined_all_pathway_frame")) {
      combined_all_pathway_frame <- currentpath 
      } else {
      combined_all_pathway_frame <- rbind.data.frame(combined_all_pathway_frame, currentpath) }
    
    if(!exists("combined_all_frame")){
        combined_all_frame <- currentframe
      }else{
        combined_all_frame <- rbind.data.frame(combined_all_frame, currentframe) }

    
  }
  
  ##pathway enrichment for sex chromosome/gonad effect 
  for(j in 1:length(resultlist_sexchrom)) {
    suppressWarnings(rm(currentframe, currentcondition))
    currentframe <- resultlist_sexchrom[[j]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    currentframe <- currentframe %>%
      dplyr::filter(abs(logFC) > 0.1 & logFC > 0) %>%
      dplyr::arrange(adj.P.Val)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcondition <- names(resultlist_sexchrom[j])
    currentframe$cell <- currentcell
    currentframe$condition <- currentcondition
    currentframe$sign <- "Upregulated"
    
    topgenes[[currentcondition]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))] ##take at most 3 
    
    sig_genes_df = subset(currentframe, adj.P.Val < 0.05 & logFC > 0) 
    genes <- sig_genes_df$logFC
    names(genes) <- sig_genes_df$gene
    genes <- na.omit(genes)
    genes <- names(genes)
    ##genes <- names(genes)[abs(genes) > 2]
    
    go_enrich <- enrichGO(gene = genes,
                      universe = gene_list,
                      OrgDb = organism, 
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)
    
    #enrichlist[[currentcondition]][[currentcell]][["UP"]] <- go_enrich_up
    
    if(is.null(go_enrich)) {next}
    
    fold <- NULL
    for(r in 1:nrow(go_enrich@result)){
      enriched_genes <- unlist(strsplit(go_enrich@result$geneID[r],"/"))
      enriched_genes <- enriched_genes[enriched_genes %in% currentframe$gene]
      fold[r] <- median(currentframe[currentframe$gene %in% enriched_genes, "logFC"], na.rm = T) }
    
    #currentpath <- go_enrich_up@result %>% dplyr::filter(p.adjust < 0.05 & qvalue < 0.05)
    currentpath <- go_enrich@result
    currentpath$fold <- fold 
    currentpath$condition <- currentcondition
    currentpath$cell <- currentcell
    currentpath$sign <- "Upregulated"
    
    #save(go_enrich_up, go_enrich_down, file = paste0("/u/home/j/julianas/project-xyang123/singlecell_cerebellum/Markers/SoupX/pathway/saves/", currentcondition, currentcell, ".Rda"))

    if(!exists("combined_all_pathway_frame")) {
      combined_all_pathway_frame <- currentpath 
      } else {
      combined_all_pathway_frame <- rbind.data.frame(combined_all_pathway_frame, currentpath) }
    
    if(!exists("combined_all_frame")){
        combined_all_frame <- currentframe
      }else{
        combined_all_frame <- rbind.data.frame(combined_all_frame, currentframe) }

  }
}

combined_all_frame_up <- combined_all_frame
combined_all_pathway_frame_up <- combined_all_pathway_frame

##Downregulated DEGs only
suppressWarnings(rm(combined_all_frame, combined_all_pathway_frame))

##Test loop 
for(i in 1:length(allfiles)) {
  currentfile <- allfiles[i]
  currentcell <- gsub(".rda","",unlist(currentfile),fixed = T)
  load(file.path("/u/project/xyang123/julianas/singlecell_cerebellum/Markers/SoupX/DEG/no_contrasts/cell_type2", currentfile))
  
  gene_list <- rownames(seurat@assays$RNA@counts)
  gene_list <- na.omit(gene_list) ##remove lowly expressed genes? 
  
  ##pathway enrichment for additionalX effect 
  for(j in 1:length(resultlist_additionalX)) {
    suppressWarnings(rm(currentframe, currentcondition))
    currentframe <- resultlist_additionalX[[j]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    currentframe <- currentframe %>%
      dplyr::filter(abs(logFC) > 0.1 & logFC < 0) %>%
      dplyr::arrange(adj.P.Val)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcondition <- names(resultlist_additionalX[j])
    currentframe$cell <- currentcell
    currentframe$condition <- currentcondition
    currentframe$sign <- "Downregulated"

    topgenes[[currentcondition]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))] ##take at most 3 
    
    sig_genes_df = subset(currentframe, adj.P.Val < 0.05 & logFC < 0) 
    genes <- sig_genes_df$logFC
    names(genes) <- sig_genes_df$gene
    genes <- na.omit(genes)
    genes <- names(genes)
    ##genes <- names(genes)[abs(genes) > 2]
    
    go_enrich <- enrichGO(gene = genes,
                      universe = gene_list,
                      OrgDb = organism, 
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)
    
    #enrichlist[[currentcondition]][[currentcell]][["UP"]] <- go_enrich_up
    
    if(is.null(go_enrich)) {next}
    
    fold <- NULL
    for(r in 1:nrow(go_enrich@result)){
      enriched_genes <- unlist(strsplit(go_enrich@result$geneID[r],"/"))
      enriched_genes <- enriched_genes[enriched_genes %in% currentframe$gene]
      fold[r] <- median(currentframe[currentframe$gene %in% enriched_genes, "logFC"], na.rm = T) }
    
    currentpath <- go_enrich@result
    currentpath$fold <- fold 
    currentpath$condition <- currentcondition
    currentpath$cell <- currentcell
    currentpath$sign <- "Downregulated"

    if(!exists("combined_all_pathway_frame")) {
      combined_all_pathway_frame <- currentpath 
      } else {
      combined_all_pathway_frame <- rbind.data.frame(combined_all_pathway_frame, currentpath) }
    
    if(!exists("combined_all_frame")){
        combined_all_frame <- currentframe
      }else{
        combined_all_frame <- rbind.data.frame(combined_all_frame, currentframe) }

    
  }
  
  ##pathway enrichment for additionalY effect 
  for(j in 1:length(resultlist_additionalY)) {
    suppressWarnings(rm(currentframe, currentcondition))
    currentframe <- resultlist_additionalY[[j]]
    if(nrow(currentframe) == 0) {next}
    
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    currentframe <- currentframe %>%
      dplyr::filter(abs(logFC) > 0.1 & logFC < 0) %>%
      dplyr::arrange(adj.P.Val)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcondition <- names(resultlist_additionalY[j])
    currentframe$cell <- currentcell
    currentframe$condition <- currentcondition
    currentframe$sign <- "Downregulated"
    
    topgenes[[currentcondition]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))] ##take at most 3 
    
    sig_genes_df = subset(currentframe, adj.P.Val < 0.05 & logFC < 0) 
    genes <- sig_genes_df$logFC
    names(genes) <- sig_genes_df$gene
    genes <- na.omit(genes)
    genes <- names(genes)
    ##genes <- names(genes)[abs(genes) > 2]
    
    go_enrich <- enrichGO(gene = genes,
                      universe = gene_list,
                      OrgDb = organism, 
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)
    
    #enrichlist[[currentcondition]][[currentcell]][["UP"]] <- go_enrich_up
    
    if(is.null(go_enrich)) {next}
    
    fold <- NULL
    for(r in 1:nrow(go_enrich@result)){
      enriched_genes <- unlist(strsplit(go_enrich@result$geneID[r],"/"))
      enriched_genes <- enriched_genes[enriched_genes %in% currentframe$gene]
      fold[r] <- median(currentframe[currentframe$gene %in% enriched_genes, "logFC"], na.rm = T) }
    
    #currentpath <- go_enrich_up@result %>% dplyr::filter(p.adjust < 0.05 & qvalue < 0.05)
    currentpath <- go_enrich@result
    currentpath$fold <- fold 
    currentpath$condition <- currentcondition
    currentpath$cell <- currentcell
    currentpath$sign <- "Downregulated"
    
    #save(go_enrich_up, go_enrich_down, file = paste0("/u/home/j/julianas/project-xyang123/singlecell_cerebellum/Markers/SoupX/pathway/saves/", currentcondition, currentcell, ".Rda"))

    if(!exists("combined_all_pathway_frame")) {
      combined_all_pathway_frame <- currentpath 
      } else {
      combined_all_pathway_frame <- rbind.data.frame(combined_all_pathway_frame, currentpath) }
    
    if(!exists("combined_all_frame")){
        combined_all_frame <- currentframe
      }else{
        combined_all_frame <- rbind.data.frame(combined_all_frame, currentframe) }

    
  }
  
  ##pathway enrichment for sex chromosome/gonad effect 
   ##pathway enrichment for additionalX effect 
  for(j in 1:length(resultlist_sexchrom)) {
    suppressWarnings(rm(currentframe, currentcondition))
    currentframe <- resultlist_sexchrom[[j]]
    if(nrow(currentframe) == 0) {next}
    s
    currentframe <- tibble::rownames_to_column(currentframe, "gene")
    currentframe <- currentframe %>%
      dplyr::filter(abs(logFC) > 0.1 & logFC < 0) %>%
      dplyr::arrange(adj.P.Val)
    currentframe <- currentframe %>% dplyr::filter(!grepl("Gm42418", gene))
    if(nrow(currentframe) == 0) {next}
    currentcondition <- names(resultlist_sexchrom[j])
    currentframe$cell <- currentcell
    currentframe$condition <- currentcondition
    currentframe$sign <- "Downregulated"
    
    topgenes[[currentcondition]][[currentcell]] <- currentframe$gene[c(1:min(3,nrow(currentframe)))] ##take at most 3 
    
   sig_genes_df = subset(currentframe, adj.P.Val < 0.05 & logFC < 0) 
    genes <- sig_genes_df$logFC
    names(genes) <- sig_genes_df$gene
    genes <- na.omit(genes)
    genes <- names(genes)
    ##genes <- names(genes)[abs(genes) > 2]
    
    go_enrich <- enrichGO(gene = genes,
                      universe = gene_list,
                      OrgDb = organism, 
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)
    
    #enrichlist[[currentcondition]][[currentcell]][["UP"]] <- go_enrich_up
    
    if(is.null(go_enrich)) {next}
    
    fold <- NULL
    for(r in 1:nrow(go_enrich@result)){
      enriched_genes <- unlist(strsplit(go_enrich@result$geneID[r],"/"))
      enriched_genes <- enriched_genes[enriched_genes %in% currentframe$gene]
      fold[r] <- median(currentframe[currentframe$gene %in% enriched_genes, "logFC"], na.rm = T) }
    
    #currentpath <- go_enrich_up@result %>% dplyr::filter(p.adjust < 0.05 & qvalue < 0.05)
    currentpath <- go_enrich@result
    currentpath$fold <- fold 
    currentpath$condition <- currentcondition
    currentpath$cell <- currentcell
    currentpath$sign <- "Downregulated"
    
    #save(go_enrich_up, go_enrich_down, file = paste0("/u/home/j/julianas/project-xyang123/singlecell_cerebellum/Markers/SoupX/pathway/saves/", currentcondition, currentcell, ".Rda"))

    if(!exists("combined_all_pathway_frame")) {
      combined_all_pathway_frame <- currentpath 
      } else {
      combined_all_pathway_frame <- rbind.data.frame(combined_all_pathway_frame, currentpath) }
    
    if(!exists("combined_all_frame")){
        combined_all_frame <- currentframe
      }else{
        combined_all_frame <- rbind.data.frame(combined_all_frame, currentframe) }

    
  }
}

combined_all_frame_down <- combined_all_frame
combined_all_pathway_frame_down <- combined_all_pathway_frame

#save(combined_all_frame, combined_all_pathway_frame, file = "Markers/SoupX/pathway/result_frames.Rda")

```

############################################################

##Visualization 

```{r}

save(combined_all_frame_up, combined_all_frame_down, combined_all_pathway_frame_up, combined_all_pathway_frame_down, topgenes, file = "Markers/SoupX/pathway/saves/celltype_normative_results.Rda")

#load(file = "Markers/SoupX/pathway/saves/celltype_results.Rda")
#load(file = "Markers/SoupX/pathway/saves/celltype2_results.Rda")

#load(file = "Markers/SoupX/pathway/saves/celltype_additionalY_ordered_model_results.Rda")
```

```{r}
load(file = "Markers/SoupX/pathway/saves/celltype_results.Rda")
combined_all_pathway_frame <- rbind(combined_all_pathway_frame_up, combined_all_pathway_frame_down)
combined_all_frame <- rbind(combined_all_frame_up, combined_all_frame_down)

combined_all_pathway_frame <- combined_all_pathway_frame %>% dplyr::filter(condition %in% c("additionalX", "Genotype_for_addXXYF", "GonadFemale", "Sex_ChromosomeXX"))
combined_all_frame <- combined_all_frame %>% dplyr::filter(condition %in% c("additionalX", "Genotype_for_addXXYF", "GonadFemale", "Sex_ChromosomeXX"))
topgenes_original <- topgenes

rm(combined_all_pathway_frame_up, combined_all_pathway_frame_down, combined_all_frame_up, combined_all_frame_down, topgenes)


##additionalY
load(file = "Markers/SoupX/pathway/saves/celltype_additionalY_ordered_model_results.Rda")
additionalY_all_pathway_frame <- rbind(combined_all_pathway_frame_up, combined_all_pathway_frame_down)
additionalY_all_frame <- rbind(combined_all_frame_up, combined_all_frame_down)
combined_all_pathway_frame <- rbind(combined_all_pathway_frame, additionalY_all_pathway_frame)
combined_all_frame <- rbind(combined_all_frame, additionalY_all_frame)

topgenes_additionalY <- topgenes 

rm(combined_all_pathway_frame_up, combined_all_pathway_frame_down, combined_all_frame_up, combined_all_frame_down, additionalY_all_pathway_frame, additionalY_all_frame, topgenes)


load(file = "Markers/SoupX/pathway/saves/celltype_additionalY_0_1_results.Rda")
additionalY_all_pathway_frame <- rbind(combined_all_pathway_frame_up, combined_all_pathway_frame_down)
additionalY_all_frame <- rbind(combined_all_frame_up, combined_all_frame_down)
combined_all_pathway_frame <- rbind(combined_all_pathway_frame, additionalY_all_pathway_frame)
combined_all_frame <- rbind(combined_all_frame, additionalY_all_frame)

topgenes_additionalY1 <- topgenes 

rm(combined_all_pathway_frame_up, combined_all_pathway_frame_down, combined_all_frame_up, combined_all_frame_down, additionalY_all_pathway_frame, additionalY_all_frame, topgenes)

load(file = "Markers/SoupX/pathway/saves/celltype_additionalY_1_2_results.Rda")
additionalY_all_pathway_frame <- rbind(combined_all_pathway_frame_up, combined_all_pathway_frame_down)
additionalY_all_frame <- rbind(combined_all_frame_up, combined_all_frame_down)
combined_all_pathway_frame <- rbind(combined_all_pathway_frame, additionalY_all_pathway_frame)
combined_all_frame <- rbind(combined_all_frame, additionalY_all_frame)

topgenes_additionalY2 <- topgenes

rm(combined_all_pathway_frame_up, combined_all_pathway_frame_down, combined_all_frame_up, combined_all_frame_down, additionalY_all_pathway_frame, additionalY_all_frame, topgenes)

##normative results
load(file = "Markers/SoupX/pathway/saves/celltype_normative_results.Rda")
normative_all_pathway_frame <- rbind(combined_all_pathway_frame_up, combined_all_pathway_frame_down)
normative_all_frame <- rbind(combined_all_frame_up, combined_all_frame_down)
combined_all_pathway_frame <- rbind(combined_all_pathway_frame, normative_all_pathway_frame)
combined_all_frame <- rbind(combined_all_frame, normative_all_frame)

combined_all_pathway_frame <- combined_all_pathway_frame %>%
  mutate(condition = replace(condition, condition == 'GenotypeXXF', 'XXF_vs_XYM'))
combined_all_frame <- combined_all_frame %>%
  mutate(condition = replace(condition, condition == 'GenotypeXXF', 'XXF_vs_XYM'))

topgenes_normative <- topgenes
rm(combined_all_pathway_frame_up, combined_all_pathway_frame_down, combined_all_frame_up, combined_all_frame_down, normative_all_pathway_frame, normative_all_frame, topgenes)

table(combined_all_frame$condition)
```

```{r}
library(UpSetR)
library(tidyr)

#gene_list <- split(combined_all_frame, f = combined_all_frame$condition)
#gene_list <- gene_list[c("additionalX", "additionalY", "Sex_ChromosomeXX", "GonadFemale")]

gene_list <- combined_all_frame %>% dplyr::filter(condition == c("XXF_vs_XYM", "additionalX", "additionalY1", "additionalY2", "Sex_ChromosomeXX", "GonadFemale")) %>% dplyr::select(gene, condition) 
gene_list <- as.data.frame(table(gene_list$gene, gene_list$condition))
gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq)
gene_list <- as.data.frame(gene_list)
gene_list[2:7] <- ifelse(gene_list[2:7] == 0, 0, 1)

#row.names(gene_list) <- gene_list$Var1
#gene_list <- gene_list[-1]
#colnames(gene_list[1]) <- "gene"
#gene_list <- ifelse(gene_list == 0, 0, 1)
#for(i in 1:ncol(gene_list)){ gene_list[ , i] <- as.integer(gene_list[ , i]) }
#gene_list <- as.matrix(gene_list)

set_vars <- c("XXF_vs_XYM", "additionalX", "additionalY1", "additionalY2", "GonadFemale", "Sex_ChromosomeXX")
plot <- UpSetR::upset(data = gene_list, sets = set_vars) ##genes filtered lfc = 0.05,p.value = 0.05 when saving DEG df's 
plot

pdf("Markers/SoupX/pathway/cell_type/upsetplot_0_1_2_model.pdf",width  = 10, height = 5)
print(plot)
dev.off()

```

```{r}
combined_all_pathway_frame <- combined_all_pathway_frame %>%
  dplyr::filter(Count >= 3 & p.adjust < 0.05)

barplotframe <- as.data.frame(table(combined_all_frame$condition, combined_all_frame$cell))
barplotframe <- barplotframe[barplotframe$Var1 %in% c("additionalX","additionalY1", "additionalY2", "GonadFemale","Sex_ChromosomeXX", "XXF_vs_XYM"),]
barplotframe$Var1 <- as.character(barplotframe$Var1)

pdf("Markers/SoupX/pathway/cell_type/additionalY_0_1_2_model/DEG_bar.pdf",width = 15,height = 5)
print(ggplot(barplotframe, aes(Var2, Freq)) + facet_grid(.~Var1,scales = "free", space = "free")+   
  geom_bar( position = "dodge", stat="identity")+xlab("")+ylab("")+
  theme_bw()+theme(axis.text.y = element_text(size = 16),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 13),
                   strip.text.x = element_text(size = 15),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"))
  )

##why these steps?
combined_all_frame$FDR <- -log10(combined_all_frame$adj.P.Val)
combined_all_frame$FDR[combined_all_frame$FDR > 3] <- 3
combined_all_frame$logFC2 <- combined_all_frame$logFC
combined_all_frame$logFC2[combined_all_frame$logFC2 > 3] <- 3
combined_all_frame$logFC2[combined_all_frame$logFC2 < -3] <- -3

combined_all_pathway_frame$FDR <- -log10(combined_all_pathway_frame$p.adjust)
combined_all_pathway_frame$FDR[combined_all_pathway_frame$FDR > 3] <- 3
combined_all_pathway_frame$fold[combined_all_pathway_frame$fold > 1] <- 1
combined_all_pathway_frame$fold[combined_all_pathway_frame$fold < -1] <- -1

#combined_all_pathway_frame$Term <- gsub(" \\(.*\\)","",combined_all_pathway_frame$Term)
genesofinterest <- c("Xist", "Eif2s3x", "Ddx3x", "Kdm5c", "Kdm6a", "Uty", "Kdm5d", "Ddx3y", "Eif2s3y")
```


PATHWAY CLEANUP - come back to this 
```{r}
dataframe <- combined_all_pathway_frame %>% dplyr::filter(condition == "additionalX") 
dataframe <- combined_all_pathway_frame %>% dplyr::filter(condition == "additionalY1")
dataframe <- combined_all_pathway_frame %>% dplyr::filter(condition == "additionalY2")
dataframe <- combined_all_pathway_frame %>% dplyr::filter(condition == "GonadFemale")
dataframe <- combined_all_pathway_frame %>% dplyr::filter(condition == "Sex_ChromosomeXX")
dataframe <- combined_all_pathway_frame %>% dplyr::filter(condition == "GenotypeXXF")

##with simplifyEnrichment -- use this for rest of pathways 
mat <- GO_similarity(dataframe$ID)
simplifyGO(mat)
df <- simplifyEnrichment(mat)
test <- left_join(dataframe, df, by=c("ID" = "id"))

additionalXframe_pathway <- test %>%
  group_by(cell, cluster) %>%
  slice_min(p.adjust) %>%
  slice_max(Count)

additionalYframe_pathway <- rbind(additionalY1frame_pathway, additionalY2frame_pathway)

table(additionalY1frame_pathway$cluster)
normativeframe_pathway %>% dplyr::filter(cluster == "3")

##manual cleanup 
additionalYframe_pathway <- additionalYframe_pathway %>% dplyr::filter(!ID == "GO:1990090")
normativeframe_pathway <- normativeframe_pathway %>% dplyr::filter(!ID %in% c("GO:0071294"))


save(additionalYframe_pathway, file = "Markers/SoupX/pathway/saves/celltype additionalY_ordered_model_cleaned.Rda")

load("Markers/SoupX/pathway/saves/cell_type_pathway_cleaned.Rda")
load("Markers/SoupX/pathway/saves/cell_type_normative_cleaned.Rda")
load("Markers/SoupX/pathway/saves/celltype_additionalY_ordered_model_cleaned.Rda")


```

```{r}
#####Loop
search_terms <- c("axon", "ATP", "cell junction", "calcium", "cardiac|heart", "cellular response", "dendrit", "detection of", "developmental", "exocyt|secretory", "eye", "adhesion", "negative regulation", "neuron", "neurotransmitter", "nucleoside|nucleotide", "transport", "presynapse", "protein", "action potential", "GTPase", "microtubule", "RNA", "vesicle")

search_df <- data.frame()
for (term in search_terms) {
  temp_df <- additionalXframe_pathway %>%
    group_by(cell) %>%
    relocate(cell, geneID) %>%
    filter(str_detect(Description, term))
  
  search_df <- rbind(search_df, temp_df)

}
search_df <- search_df %>% distinct()
search_df$cell <- factor(search_df$cell)

additionalXframe_pathway_unique <- additionalXframe_pathway[!additionalXframe_pathway$ID %in% search_df$ID, ]
additionalXframe_pathway_cleaned <- list()
i = 1
for (term in search_terms) {
  temp_df <- search_df %>%
    group_by(cell) %>%
    relocate(cell, geneID, Description) %>%
    dplyr::filter(str_detect(Description, term)) %>%
    slice_min(p.adjust, with_ties = FALSE)
  
  additionalXframe_pathway_cleaned[[i]] <- temp_df
  i = i+1
  
}
df <- do.call("rbind", additionalXframe_pathway_cleaned)

save(additionalXframe_pathway_unique, df, file = "Markers/SoupX/pathway/saves/additionalX_pathway_cleanup_frames_loop.Rda")
additionalXframe_pathway <- rbind(additionalXframe_pathway_unique, df)


```

```{r}
##manual pathway cleanup 
original <- search_df %>%
  dplyr::group_by(cell) %>%
  dplyr::relocate(cell, geneID, Description) %>%
  dplyr::filter(str_detect(Description, "vesicle") & !str_detect(Description, "exocytosis"))

query <- search_df %>%
  group_by(cell) %>%
  relocate(cell, geneID, Description) %>%
  dplyr::filter(str_detect(Description, "vesicle") & !str_detect(Description, "exocytosis")) %>%
  slice_min(p.adjust, with_ties = FALSE)

extra_rows <- search_df[(search_df$ID %in% c("GO:0019228")), ]
extra_rows <- extra_rows[!(extra_rows$ID %in% c("GO:0034764") & extra_rows$cell == "MLI1"), ]
rows_to_remove <- query[query$ID %in% c("GO:0034764"), ]

original
query
extra_rows
rows_to_remove

query <- rbind(query, extra_rows)
query <- query[!query$ID %in% rows_to_remove$ID, ]

additionalXframe_pathway_cleaned <- rbind(additionalXframe_pathway_cleaned, query)

additionalXframe_pathway <- rbind(additionalXframe_pathway_unique, df)

rows_to_remove <- additionalXframe_pathway %>%
  group_by(cell) %>%
  relocate(cell, geneID, Description) %>%
  dplyr::filter(str_detect(Description, "transport"))
rows_to_add <- additionalXframe_pathway %>%
  group_by(cell) %>%
  relocate(cell, geneID, Description) %>%
  dplyr::filter(str_detect(Description, "transport")) %>%
  slice_min(p.adjust, with_ties = FALSE)
extra_rows <- rows_to_remove[(rows_to_remove$ID %in% c("GO:0043270") & rows_to_remove$cell == "Granule"), ]

rows_to_remove
rows_to_add
extra_rows

additionalXframe_pathway <- additionalXframe_pathway[!additionalXframe_pathway$ID %in% rows_to_remove$ID, ]
additionalXframe_pathway <- rbind(additionalXframe_pathway, rows_to_add)

head(rows_to_remove)
```

```{r}
additionalXframe <- combined_all_frame %>% dplyr::filter(condition == "additionalX") 
genedots <- unique(unlist(topgenes_original$additionalX))
test <- as.data.frame(table(additionalXframe$gene))
test <- test[order(test$Freq,decreasing = T),]
uniquegene <- additionalXframe[additionalXframe$gene %in% test$Var1[test$Freq == 1],]
uniquegene <- uniquegene %>% group_by(cell) %>% slice_min(order_by = adj.P.Val, n = 3) 

test <- as.data.frame(table(additionalXframe_pathway$Description))
test <- test[order(test$Freq,decreasing = T),]
selectedpathway <- as.character(test$Var1[test$Freq %in% c(max(test$Freq), max(test$Freq)-1)]) ##edit this line to take top couple? 
uniquepathway <- additionalXframe_pathway[additionalXframe_pathway$Description %in% test$Var1[test$Freq == 1],]
uniquepathway <- uniquepathway %>% group_by(cell) %>% slice_min(order_by = p.adjust, n = 3)

p1 <- ggplot(additionalXframe[additionalXframe$gene %in% genedots,], aes(y = gene,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("additionalX effect")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 18),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))
p1_3 <- ggplot(additionalXframe[additionalXframe$gene %in% genesofinterest,], aes(y = gene,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("additionalX effect")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 18),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))

p1_2 <- ggplot(uniquegene, aes(y = factor(gene,levels = uniquegene$gene),x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("additionalX effect-unique top DEG")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 18),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))

p2 <- ggplot(additionalXframe_pathway[additionalXframe_pathway$Description %in% selectedpathway,], aes(y = Description,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("additionalX effect")+      ## to get the rect filled
  geom_point(aes(colour = fold, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-1,1),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 18),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))
p2_2 <- ggplot(uniquepathway, aes(y = factor(Description,levels = uniquepathway$Description),x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("additionalX effect-unique top pathway")+      ## to get the rect filled
  geom_point(aes(colour = fold, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-1,1),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 18),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))

pdf("Markers/SoupX/pathway/cell_type/original/additionalX.pdf",width  = 25, height = 20)
print(grid.arrange(p1,p1_2,p2,p2_2,nrow = 2))
dev.off()
```

```{r}
ggplot(additionalXframe[additionalXframe$gene %in% genesofinterest, ], aes(y = gene,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("additionalX effect")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text(face = "italic", size = 13),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 18),
                   strip.text.x = element_text(size = 20),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))
```

```{r}
additionalYframe <- combined_all_frame[combined_all_frame$condition %in% "additionalY1",]
genedots <- unique(unlist(topgenes_additionalY$additionalY1))
test <- as.data.frame(table(additionalYframe$gene))
test <- test[order(test$Freq,decreasing = T),]
uniquegene <- additionalYframe[additionalYframe$gene %in% test$Var1[test$Freq == 2], ] ##no gene with frequency of 1 
uniquegene <- uniquegene %>% group_by(cell) %>% slice_min(order_by = adj.P.Val, n = 3)

test <- as.data.frame(table(additionalYframe_pathway$Description))
test <- test[order(test$Freq,decreasing = T),]
selectedpathway <- as.character(test$Var1[test$Freq %in% c(max(test$Freq), max(test$Freq)-2)])
uniquepathway <- additionalYframe_pathway[additionalYframe_pathway$Description %in% test$Var1[test$Freq == 1], ]
uniquepathway <- uniquepathway %>% group_by(cell) %>% slice_min(order_by = p.adjust, n = 3)


p3 <- ggplot(additionalYframe[additionalYframe$gene %in% genedots,], aes(y = gene,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("additionalY1 effect")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 18),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))
p3_3 <- ggplot(additionalYframe[additionalYframe$gene %in% genesofinterest,], aes(y = gene,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("additionalY1 effect")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 18),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))

p3_2 <- ggplot(uniquegene, aes(y = factor(gene,levels = unique(uniquegene$gene)),x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("additionalY1 effect-unique top DEG")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 18),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))


p4 <- ggplot(additionalYframe_pathway[additionalYframe_pathway$Description %in% selectedpathway,], aes(y = Description,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("additionalY1 effect")+      ## to get the rect filled
  geom_point(aes(colour = fold, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-1,1),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 18),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))

p4_2 <- ggplot(uniquepathway, aes(y = factor(Description,levels = unique(uniquepathway$Description)),x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("additionalY1 effect-unique top pathway")+      ## to get the rect filled
  geom_point(aes(colour = fold, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-1,1),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 18),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))

pdf("Markers/SoupX/pathway/cell_type/additionalY_ordered_model/additionalY1.pdf",width  = 25, height = 20)
print(grid.arrange(p3,p3_2,p4,p4_2,ncol = 2))
dev.off()
```

```{r}
additionalYframe <- combined_all_frame[combined_all_frame$condition %in% "additionalY2",]
genedots <- unique(unlist(topgenes_additionalY$additionalY2))
test <- as.data.frame(table(additionalYframe$gene))
test <- test[order(test$Freq,decreasing = T),]
uniquegene <- additionalYframe[additionalYframe$gene %in% test$Var1[test$Freq == 2], ] ##no gene with frequency of 1 
uniquegene <- uniquegene %>% group_by(cell) %>% slice_min(order_by = adj.P.Val, n = 3)

test <- as.data.frame(table(additionalYframe_pathway$Description))
test <- test[order(test$Freq,decreasing = T),]
selectedpathway <- as.character(test$Var1[test$Freq %in% c(max(test$Freq), max(test$Freq)-2)])
uniquepathway <- additionalYframe_pathway[additionalYframe_pathway$Description %in% test$Var1[test$Freq == 1], ]
uniquepathway <- uniquepathway %>% group_by(cell) %>% slice_min(order_by = p.adjust, n = 3)


p3 <- ggplot(additionalYframe[additionalYframe$gene %in% genedots,], aes(y = gene,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("additionalY2 effect")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 18),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))
p3_3 <- ggplot(additionalYframe[additionalYframe$gene %in% genesofinterest,], aes(y = gene,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("additionalY2 effect")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 18),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))

p3_2 <- ggplot(uniquegene, aes(y = factor(gene,levels = unique(uniquegene$gene)),x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("additionalY2 effect-unique top DEG")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 18),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))


p4 <- ggplot(additionalYframe_pathway[additionalYframe_pathway$Description %in% selectedpathway,], aes(y = Description,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("additionalY2 effect")+      ## to get the rect filled
  geom_point(aes(colour = fold, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-1,1),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 18),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))

p4_2 <- ggplot(uniquepathway, aes(y = factor(Description,levels = unique(uniquepathway$Description)),x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("additionalY2 effect-unique top pathway")+      ## to get the rect filled
  geom_point(aes(colour = fold, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-1,1),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 18),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))

pdf("Markers/SoupX/pathway/cell_type/additionalY_ordered_model/additionalY2.pdf",width  = 25, height = 20)
print(grid.arrange(p3,p3_2,p4,p4_2,ncol = 2))
dev.off()
```

```{r}
ggplot(additionalYframe[additionalYframe$gene %in% genesofinterest, ], aes(y = gene,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("additionalY effect")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text(face = "italic", size = 13),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 18),
                   strip.text.x = element_text(size = 20),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))
```

```{r}
sexchromeXXframe <- combined_all_frame[combined_all_frame$condition %in% "Sex_ChromosomeXX",]
genedots <- unique(unlist(topgenes_original$Sex_ChromosomeXX))
test <- as.data.frame(table(sexchromeXXframe$gene))
test <- test[order(test$Freq,decreasing = T),]
uniquegene <- sexchromeXXframe[sexchromeXXframe$gene %in% test$Var1[test$Freq == 1],]
uniquegene <- uniquegene %>% group_by(cell) %>% slice_min(order_by = adj.P.Val, n = 3)

test <- as.data.frame(table(sexchromXXframe_pathway$Description))
test <- test[order(test$Freq,decreasing = T),]
selectedpathway <- as.character(test$Var1[test$Freq %in% max(test$Freq)])
uniquepathway <- sexchromXXframe_pathway[sexchromXXframe_pathway$Description %in% test$Var1[test$Freq == 1],]
uniquepathway <- uniquepathway %>% group_by(cell) %>% slice_min(order_by = p.adjust, n = 3)

p5 <- ggplot(sexchromeXXframe[sexchromeXXframe$gene %in% genedots,], aes(y = gene,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("Sex_ChromosomeXX effect")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 18),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))

p5_3 <- ggplot(sexchromeXXframe[sexchromeXXframe$gene %in% genesofinterest,], aes(y = gene,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("Sex_ChromosomeXX effect")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 18),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))

p5_2 <- ggplot(uniquegene, aes(y = factor(gene,levels = uniquegene$gene),x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("Sex_ChromosomeXX effect-unique top DEG")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 18),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))


p6 <- ggplot(sexchromXXframe_pathway[sexchromXXframe_pathway$Description %in% selectedpathway,], aes(y = Description,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("Sex_ChromosomeXX effect")+      ## to get the rect filled
  geom_point(aes(colour = fold, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-1,1),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 18),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))

p6_2 <- ggplot(uniquepathway, aes(y = factor(Description,levels = uniquepathway$Description),x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("Sex_ChromosomeXX effect-unique top pathway")+      ## to get the rect filled
  geom_point(aes(colour = fold, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-1,1),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 18),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
                   strip.text.x = element_text(size = 25),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"),
                   plot.title = element_text(size = 25))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))


pdf("Markers/SoupX/pathway/cell_type/original/sexchromXX.pdf",width  = 25, height = 20)
print(grid.arrange(p5,p5_2,p6,p6_2,ncol = 2))
dev.off()
```

```{r}
ggplot(sexchromeXXframe[sexchromeXXframe$gene %in% genesofinterest, ], aes(y = gene,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("Sex Chromosome XX effect")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text(face = "italic", size = 13),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 18),
                   strip.text.x = element_text(size = 20),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))
```

```{r}
GonadFemaleframe <- combined_all_frame[combined_all_frame$condition %in% "GonadFemale",]
genedots <- unique(unlist(topgenes$Sex_ChromosomeXX))
test <- as.data.frame(table(GonadFemaleframe$gene))
test <- test[order(test$Freq,decreasing = T),]
uniquegene <- GonadFemaleframe[GonadFemaleframe$gene %in% test$Var1[test$Freq == 1],]
uniquegene <- uniquegene %>% group_by(cell) %>% slice_min(order_by = adj.P.Val, n = 3)

test <- as.data.frame(table(GonadFemaleframe_pathway$Description))
test <- test[order(test$Freq,decreasing = T),]
selectedpathway <- as.character(test$Var1[test$Freq %in% max(test$Freq)])
uniquepathway <- GonadFemaleframe_pathway[GonadFemaleframe_pathway$Description %in% test$Var1[test$Freq == 1],]
uniquepathway <- uniquepathway %>% group_by(cell) %>% slice_min(order_by = p.adjust, n = 3)

p7 <- ggplot(GonadFemaleframe[GonadFemaleframe$gene %in% genedots,], aes(y = gene,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("GonadFemale effect")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text(face = "italic", size = 13),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 18),
                   strip.text.x = element_text(size = 20),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))

p7_3 <- ggplot(GonadFemaleframe[GonadFemaleframe$gene %in% genesofinterest,], aes(y = gene,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("GonadFemaleframe effect")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text(face = "italic", size = 13),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 18),
                   strip.text.x = element_text(size = 20),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))

p7_2 <- ggplot(uniquegene, aes(y = factor(gene,levels = uniquegene$gene),x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("GonadFemale effect-unique top DEG")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text(face = "italic", size = 13),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 18),
                   strip.text.x = element_text(size = 20),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))


p8 <- ggplot(GonadFemaleframe_pathway[GonadFemaleframe_pathway$Description %in% selectedpathway,], aes(y = Description,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("GonadFemale effect")+      ## to get the rect filled
  geom_point(aes(colour = fold, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-1,1),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text(face = "italic", size = 13),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 18),
                   strip.text.x = element_text(size = 20),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))
p8_2 <- ggplot(uniquepathway, aes(y = factor(Description,levels = uniquepathway$Description),x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("GonadFemale effect-unique top pathway")+      ## to get the rect filled
  geom_point(aes(colour = fold, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-1,1),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 13),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 18),
                   strip.text.x = element_text(size = 20),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))




pdf("Markers/SoupX/pathway/cell_type2/GonadF.pdf",width  = 25, height = 25)
print(grid.arrange(p7,p7_2,p8,p8_2,ncol = 2))
dev.off()
```

```{r}
ggplot(GonadFemaleframe[GonadFemaleframe$gene %in% genesofinterest, ], aes(y = gene,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("Gonad F effect")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text(face = "italic", size = 13),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 18),
                   strip.text.x = element_text(size = 20),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))
```

```{r}
normativeframe <- combined_all_frame %>% dplyr::filter(condition == "GenotypeXXF") 
genedots <- unique(unlist(topgenes$GenotypeXXF))
test <- as.data.frame(table(normativeframe$gene))
test <- test[order(test$Freq,decreasing = T),]
uniquegene <- normativeframe[normativeframe$gene %in% test$Var1[test$Freq == 1],]
uniquegene <- uniquegene %>% group_by(cell) %>% slice_min(order_by = adj.P.Val, n = 3) 

test <- as.data.frame(table(normativeframe_pathway$Description))
test <- test[order(test$Freq,decreasing = T),]
selectedpathway <- as.character(test$Var1[test$Freq %in% c(max(test$Freq), max(test$Freq)-1, max(test$Freq)-2)]) ##edit this line to take top couple? 
uniquepathway <- normativeframe_pathway[normativeframe_pathway$Description %in% test$Var1[test$Freq == 1],]
uniquepathway <- uniquepathway %>% group_by(cell) %>% slice_min(order_by = p.adjust, n = 3)

p9 <- ggplot(normativeframe[normativeframe$gene %in% genedots,], aes(y = gene,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("Normative difference")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text(face = "italic", size = 13),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 18),
                   strip.text.x = element_text(size = 20),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))
p9_3 <- ggplot(normativeframe[normativeframe$gene %in% genesofinterest,], aes(y = gene,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("Normative difference")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text(face = "italic", size = 13),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 18),
                   strip.text.x = element_text(size = 20),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))

p9_2 <- ggplot(uniquegene, aes(y = factor(gene,levels = uniquegene$gene),x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("Normative difference-unique top DEG")+      ## to get the rect filled
  geom_point(aes(colour = logFC2, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-3,3),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text(face = "italic", size = 13),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 18),
                   strip.text.x = element_text(size = 20),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))

p10 <- ggplot(normativeframe_pathway[normativeframe_pathway$Description %in% selectedpathway,], aes(y = Description,x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("Normative difference")+      ## to get the rect filled
  geom_point(aes(colour = fold, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-1,1),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text(size = 13),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 18),
                   strip.text.x = element_text(size = 20),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))

p10_2 <- ggplot(uniquepathway, aes(y = factor(Description,levels = uniquepathway$Description),x = cell)) +
  geom_tile(fill = "white") +  xlab("")+ylab("")+ ggtitle("Normative difference-unique top pathway")+      ## to get the rect filled
  geom_point(aes(colour = fold, 
                 size = FDR))  +    ## geom_point for circle illusion
  scale_color_gradientn(limits = c(-1,1),colours  = colorRampPalette(rev(brewer.pal(11,"RdBu")))(40))+   ## color of the corresponding aes          ## to tune the size of circles
  theme_bw()+theme(axis.text.y = element_text( size = 13),
                   axis.text.x = element_text(angle = 45, hjust = 1, size = 18),
                   strip.text.x = element_text(size = 20),
                   plot.margin = unit(c(0.5,1,0.5,1.3), "cm"))+ guides(colour=guide_legend(title="-log2(fold change)"),
                                                                       size =guide_legend(title="-log10(FDR)"))

pdf("Markers/SoupX/pathway/cell_type/normative/normative.pdf",width  = 25, height = 25)
print(grid.arrange(p9,p9_2,p10,p10_2,nrow = 2))
dev.off()
```
