---
title: "overlap_analysis"
output: html_document
date: "2024-04-25"
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/u/project/xyang123/jshin/medial_septum_sct")

# library(limma)
# library(edgeR)
# library(data.table)
library(plyr)
library(dplyr)
library(Matrix)
library(Seurat)
#library(enrichplot)
#library(org.Mm.eg.db)
#library(ggplot2)
#library(RColorBrewer)
#library(grid)
#library(gridExtra)
#library(Mus.musculus)
library(MAST)
library(lme4)
library(survcomp)
library(ggplot2)
library(UpSetR)
library(tidyverse)
library(ComplexUpset)
library(ggpubr)
library(readxl)
library(GeneOverlap)
```

##DEG datasets
```{r}
load(file = "Results/DEG_processed/DEG_v2/DEG_df.rda")

gene_annotation <- readRDS("Gene_annotation/gene_annotation.Rds")
gene_annotation2 <- gene_annotation[!duplicated(gene_annotation$mgi_symbol), c(1:3, 7:9)]
gene_list <- read.table("Gene_annotation/gene_universe.txt") %>% unlist(use.names = FALSE)

allcells <- unique(DEG_df$cell_type)
gene_universe <- read.table("Gene_annotation/gene_universe.txt") %>% unlist(use.names = FALSE)

DEG_df_annotated <- left_join(DEG_df, gene_annotation2, by = c("gene"="mgi_symbol"))
DEG_df_annotated$chromosome_name <- factor(DEG_df_annotated$chromosome_name, levels = c((1:19),"X","Y","MT"))
```

##Overlap analysis - external gene sets 
```{r}
##Read in control sets 
estrogen1 <- readxl::read_xlsx("Gene_annotation/external_gene_sets/Gegenhuber_2022_oestradiol_1.xlsx", sheet = 1) %>% dplyr::filter(padj < 0.05)
colnames(estrogen1)[1] <- "gene"
estrogen2 <- readxl::read_xlsx("Gene_annotation/external_gene_sets/Gegenhuber_2022_oestradiol_3.xlsx", sheet = 1) %>% dplyr::filter(padj < 0.05)
colnames(estrogen2)[1] <- "gene"
estrogen3 <- readxl::read_xlsx("Gene_annotation/external_gene_sets/humphreys_2014_e2.xlsx", sheet = 2)
estrogen4 <- readxl::read_xlsx("Gene_annotation/external_gene_sets/gocz_2022_estrogen1.xlsx", sheet = 1, skip = 1) %>% 
  dplyr::filter(p.adj. < 0.05)
estrogen5 <- readxl::read_xlsx("Gene_annotation/external_gene_sets/gocz_2022_estrogen2.xlsx", sheet = 1, skip = 1) %>% 
  dplyr::filter(p.adj < 0.05)
# estrogen6 <- readxl::read_xlsx("Gene_annotation/external_gene_sets/fels_2021_erb.xlsx", sheet = 2) %>% 
#   dplyr::filter(padj < 0.05)

estrogen_list <- list()
for(i in 1:7) {estrogen_list[[i]] <- readxl::read_xlsx("Gene_annotation/external_gene_sets/Gegenhuber_2022_oestradiol_2.xlsx", sheet = i) %>% dplyr::filter(padj < 0.05)
}
names(estrogen_list) <- excel_sheets(path = "Gene_annotation/external_gene_sets/Gegenhuber_2022_oestradiol_2.xlsx")

li_androgen <- read_xlsx(path = "~/project-xyang123/cerebellum_sct/Gene_annotation/external_gene_sets/li_2024_androgens.xlsx", col_names = TRUE) %>%
  dplyr::filter(Tissue == "Brain" & p_val_adj < 0.05)

androgen1 <- li_androgen %>% filter(Comparison == "MSvsFS")
androgen2 <- li_androgen %>% filter(Comparison == "FDvsFS")
#androgen3 <- read_xlsx(path = "~/project-xyang123/cerebellum_sct/Gene_annotation/external_gene_sets/keleva_2022_androgen.xlsx", sheet = 2, col_names = TRUE)

x_chromosome_genes <- DEG_df_annotated %>% dplyr::filter(chromosome_name == "X")
y_chromosome_genes <- DEG_df_annotated %>% dplyr::filter(chromosome_name == "Y")
escape_genes <- DEG_df_annotated %>% dplyr::filter(Escapee == "Escapee")

hormone_control_genes <- list(estrogen1$gene, estrogen2$gene, estrogen_list[[1]]$...1, estrogen_list[[2]]$...1, estrogen_list[[3]]$...1, estrogen_list[[4]]$...1, estrogen_list[[5]]$...1, estrogen_list[[6]]$...1, estrogen_list[[7]]$...1, estrogen3$`Gene name`, estrogen4$`Gene symbol`, estrogen5$`Gene symbol`, androgen1$Gene_symbol, androgen2$Gene_symbol)
names(hormone_control_genes) <- c("gonad_estrogen_1", "gonad_estrogen_2", "gonad_estrogen_3", "gonad_estrogen_4", "gonad_estrogen_5", "gonad_estrogen_6", "gonad_estrogen_7", "gonad_estrogen_8", "gonad_estrogen_9", "gonad_estrogen_10", "gonad_estrogen_11", "gonad_estrogen_12", "gonad_androgen_1", "gonad_androgen_2")

sex_control_genes <- list(x_chromosome_genes$gene, escape_genes$gene, y_chromosome_genes$gene)
names(sex_control_genes) <- c("X_chromosome", "X_inactivation_escape", "Y_chromosome")

control_genes <- c(hormone_control_genes, sex_control_genes)
names(control_genes) <- c(names(hormone_control_genes), names(sex_control_genes))
#################################################################
cell_list <- list()
for(i in 1:length(allcells)){
  
  currentcell <- allcells[i]
  deg_list <- DEG_df %>% dplyr::filter(cell_type == currentcell) %>% split(., f = .$effect)
  deg_list <- lapply(deg_list, function(x) {x %>% dplyr::select(gene) %>% unlist(use.names = FALSE)})

  go.obj<- newGOM(control_genes, deg_list, genome.size = length(gene_universe))

  pvals <- getMatrix(go.obj, name="pval") %>% as.data.frame() %>% tibble::rownames_to_column(var = "setA") %>% pivot_longer(!setA, names_to = "setB", values_to = "pval")
  odds <- getMatrix(go.obj, name="odds.ratio") %>% as.data.frame() %>% tibble::rownames_to_column(var = "setA") %>% pivot_longer(!setA, names_to = "setB", values_to = "odds")
  
  cell_counts <- lapply(deg_list, function(x) {length(x)}) %>% do.call("rbind", .) %>% as.data.frame() %>% tibble::rownames_to_column(var = "set")
  df <- left_join(pvals, odds, by=c("setA" = "setA", "setB" = "setB"))
  df <- left_join(df, cell_counts, by=c("setB" = "set")) %>% dplyr::rename(setB_size = V1)
  df$cell_type <- currentcell
  df$adj.pval <- p.adjust(df$pval, method = "BH", n = (nrow(df)*length(allcells)))
  
  df$odds <- ifelse(df$adj.pval > 0.05, NA, df$odds)
  
  cell_list[[currentcell]] <- df
}

plot_list <- lapply(cell_list, function(x) {
  subset <- x[is.finite(x$odds), ]
  max <- max(subset$odds) %>% ceiling()
  min <- min(subset$odds) %>% floor()
  
  x$odds <- ifelse(x$odds == Inf, max + 10, x$odds)
  x$setB <- factor(x$setB, levels = c("Normative", "Gonad", "SexChrom", "Gonad:SexChrom", "addX", "addY"))
  
  ggplot(x, aes(x = setA, y = setB, fill = odds)) +
    geom_tile() +
    theme_minimal() +
    scale_fill_gradient2(low="pink", high="darkred", limits = c(0, max + 10)) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=0.95, size = 30),
          axis.text.y = element_text(size = 30),
          axis.title.x=element_blank(),
          axis.title.y=element_blank(),
          text = element_text(size= 30)) +
    ggtitle(x$cell_type)
})

pdf("Plots/Caden/DEG/overlap/control_sets_overlap.pdf", height = 20, width = 50)
print(ggarrange(plotlist = plot_list, ncol = 4, nrow = 2))
dev.off()
```

########################################################

STACKED BARPLOT OF NORMATIVE DIFFERENCES
```{r}
DEG_df_filtered <- DEG_df %>% dplyr::filter(adj.P.Val < 0.05 & abs(logFC) > 0.1 & !cell_type %in% c("Progenitor")) %>% 
  mutate(model = factor(effect, levels = c("Normative", "Gonad", "SexChrom", "Gonad:SexChrom", "addX", "addY")),
         cell_type = factor(cell_type)) %>% dplyr::select(gene, effect, cell_type)

sets <- as.data.frame(table(DEG_df_filtered$gene, DEG_df_filtered$effect, DEG_df_filtered$cell_type)) %>% 
  pivot_wider(names_from = Var2, values_from = Freq) %>%
  as.data.frame(check.names = FALSE) %>%
  dplyr::filter(!Normative == 0) %>%
  mutate(shared = ifelse(rowSums(.[, 3:8]) > 2, 1, 0)) %>% ##make sure all columns are selected here 
  group_by(Var3) %>%
  relocate(Var1, Var3, Normative)

for(col in c("Gonad", "SexChrom", "Gonad:SexChrom", "addX", "addY")) {
  sets[[col]] <- ifelse(sets$shared == 1, 0, sets[[col]])
}

sets <- sets %>% summarise(across(where(is.numeric), sum)) %>%
  mutate(normative_unique = Normative - Gonad - SexChrom - `Gonad:SexChrom` - addX - addY - shared)

sets_prop <- t(apply(sets[-1], 1, function(row) row / row[1])) %>% as.data.frame()
rownames(sets_prop) <- sets$Var3
sets_prop <- sets_prop %>% rownames_to_column(var = "cell_type")
sets_prop$Normative <- NULL
sets_prop <- sets_prop %>% 
  pivot_longer(cols = !cell_type, names_to = "Set", values_to = "Proportion")
sets_prop$Set <- factor(sets_prop$Set, levels = c("addY", "addX", "Gonad:SexChrom", "SexChrom", "Gonad", "shared", "normative_unique"))

ggplot(sets_prop, aes(x = cell_type, y = Proportion, fill = Set)) +
  geom_bar(stat = "identity") +
  labs(
    x = "",
    y = "Proportion",
    fill = "Category"
  ) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=0.95),
        text = element_text(size = 14)) +
  scale_fill_manual(values=c("#EEA9B8","#90EE90", "#698B69", "#68228B", "#E69F00", "#1874CD", "#999999"))

pdf("Plots/Caden/Barplot/Normative_overlap_proportion.pdf", width  = 5,  height = 8)
print(last_plot())
dev.off()

### ### ### ### ### ### ### ### ### ### 
##FCG overlap

DEG_df_filtered <- DEG_df %>% dplyr::filter(adj.P.Val < 0.05 & abs(logFC) > 0.1 & !cell_type %in% c("Progenitor") & effect %in% c("Normative", "Gonad", "SexChrom", "Gonad:SexChrom")) %>% 
  mutate(model = factor(effect, levels = c("Normative", "Gonad", "SexChrom", "Gonad:SexChrom")),
         cell_type = factor(cell_type)) %>% dplyr::select(gene, effect, cell_type)

sets <- as.data.frame(table(DEG_df_filtered$gene, DEG_df_filtered$effect, DEG_df_filtered$cell_type)) %>% 
  pivot_wider(names_from = Var2, values_from = Freq) %>%
  as.data.frame(check.names = FALSE) %>%
  dplyr::filter(!Normative == 0) %>%
  mutate(shared = ifelse(rowSums(.[, 3:6]) > 2, 1, 0)) %>% ##make sure all columns are selected here 
  group_by(Var3) %>%
  relocate(Var1, Var3, Normative)

for(col in c("Gonad", "SexChrom", "Gonad:SexChrom")) {
  sets[[col]] <- ifelse(sets$shared == 1, 0, sets[[col]])
}

sets <- sets %>% summarise(across(where(is.numeric), sum)) %>%
  mutate(normative_unique = Normative - Gonad - SexChrom - `Gonad:SexChrom` - shared)

sets_prop <- t(apply(sets[-1], 1, function(row) row / row[1])) %>% as.data.frame()
rownames(sets_prop) <- sets$Var3
sets_prop <- sets_prop %>% rownames_to_column(var = "cell_type")
sets_prop$Normative <- NULL
sets_prop <- sets_prop %>% 
  pivot_longer(cols = !cell_type, names_to = "Set", values_to = "Proportion")
sets_prop$Set <- factor(sets_prop$Set, levels = c("Gonad:SexChrom", "SexChrom", "Gonad", "shared", "normative_unique"))

ggplot(sets_prop, aes(x = cell_type, y = Proportion, fill = Set)) +
  geom_bar(stat = "identity") +
  labs(
    x = "",
    y = "Proportion",
    fill = "Category"
  ) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=0.95),
        text = element_text(size = 14)) +
  scale_fill_manual(values=c("#698B69", "#68228B", "#E69F00", "#1874CD", "#999999"))

pdf("Plots/Caden/Barplot/Normative_overlap_proportion_FCG.pdf", width  = 5,  height = 8)
print(last_plot())
dev.off()
```

##################################################

UPSET PLOTS - edit for limma results 

```{r}
## MASHR/global output 
DEG_df_filtered <- global_df %>% dplyr::filter(adj.P.Val < 0.05) %>% 
  mutate(model = factor(model, levels = c("Normative", "FCG", "Dose", "Dose2", "Aneuploidy", "Complete")),
         cell_type = factor(cell_type))

##FCG 
plot_list <- list()
for (i in 1:length(allcells)) {
 
  currentcell <- as.character(allcells[i])
  gene_list <- DEG_df_filtered %>% dplyr::filter(model %in% c("Normative", "FCG") & cell_type == currentcell) %>% dplyr::select(gene, comparison)
  gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
  gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame(check.names = FALSE)
  gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)

  set_vars <- c("XX_vs_XY", "Ovary_vs_Testis", "XXF_vs_XYM")
  gene_list <- gene_list %>% dplyr::select(Var1, all_of(set_vars))
  
  plot <- ComplexUpset::upset(gene_list, names(gene_list)[-1], sort_sets = FALSE, sort_intersections = FALSE, name = currentcell,
                              themes=upset_default_themes(text=element_text(size=30)))
  
  plot_list[[currentcell]] <- plot
}

pdf("Plots/DEG/global/upset/FCG_overlap.pdf", width  = 50, height = 30)
do.call(ggarrange, c(plot_list, list(ncol = 5, nrow = 3)))
dev.off()


##Dose 
plot_list <- list()
for (i in 1:length(allcells)) {
 
  currentcell <- as.character(allcells[i])
  gene_list <- DEG_df_filtered %>% dplyr::filter(model %in% c("Normative", "Dose") & cell_type == currentcell) %>% dplyr::select(gene, comparison)
  gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
  gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame(check.names = FALSE)
  gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)

  set_vars <- c("Xdose:Ydose", "Gonad:Ydose", "Gonad:Xdose", "Ydose", "Xdose", "Ovary_vs_Testis", "XXF_vs_XYM")
  gene_list <- gene_list %>% dplyr::select(Var1, any_of(set_vars))
  
  plot <- ComplexUpset::upset(gene_list, names(gene_list)[-1], sort_sets = FALSE, sort_intersections = FALSE, name = currentcell,
                              themes=upset_default_themes(text=element_text(size=30)))
  
  plot_list[[currentcell]] <- plot
}

pdf("Plots/DEG/global/upset/Dose_overlap.pdf", width  = 70, height = 30)
do.call(ggarrange, c(plot_list, list(ncol = 5, nrow = 3)))
dev.off()


##Dose2
plot_list <- list()
for (i in 1:length(allcells)) {
 
  currentcell <- as.character(allcells[i])
  gene_list <- DEG_df_filtered %>% dplyr::filter(model %in% c("Normative", "Dose2") & cell_type == currentcell) %>% dplyr::select(gene, comparison)
  gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
  gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame(check.names = FALSE)
  gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)

  set_vars <- c("Gonad:Ydose2", "Gonad:Ydose1", "Gonad:Xdose", "Ydose2", "Ydose1", "Xdose", "Ovary_vs_Testis", "XXF_vs_XYM")
  gene_list <- gene_list %>% dplyr::select(Var1, any_of(set_vars))
  
  plot <- ComplexUpset::upset(gene_list, names(gene_list)[-1], sort_sets = FALSE, sort_intersections = FALSE, name = currentcell,
                              themes=upset_default_themes(text=element_text(size=30)))
  
  plot_list[[currentcell]] <- plot
}

pdf("Plots/DEG/global/upset/Dose2_overlap.pdf", width  = 70, height = 30)
do.call(ggarrange, c(plot_list, list(ncol = 5, nrow = 3)))
dev.off()

##Aneuploidy
plot_list <- list()
for (i in 1:length(allcells)) {
 
  currentcell <- as.character(allcells[i])
  gene_list <- DEG_df_filtered %>% dplyr::filter(model %in% c("Normative", "Aneuploidy") & cell_type == currentcell) %>% dplyr::select(gene, comparison)
  gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
  gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame(check.names = FALSE)
  gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)

  set_vars <- c("Gonad:Aneuploidy", "Gonad:SexChrom", "Aneuploidy", "XX_vs_XY", "Ovary_vs_Testis", "XXF_vs_XYM")
  gene_list <- gene_list %>% dplyr::select(Var1, any_of(set_vars))
  
  plot <- ComplexUpset::upset(gene_list, names(gene_list)[-1], sort_sets = FALSE, sort_intersections = FALSE, name = currentcell,
                              themes=upset_default_themes(text=element_text(size=30)))
  
  plot_list[[currentcell]] <- plot
}

pdf("Plots/DEG/global/upset/Aneuploidy_overlap.pdf", width  = 70, height = 30)
do.call(ggarrange, c(plot_list, list(ncol = 5, nrow = 3)))
dev.off()

##Complete
plot_list <- list()
for (i in 1:length(allcells)) {
 
  currentcell <- as.character(allcells[i])
  gene_list <- DEG_df_filtered %>% dplyr::filter(model %in% c("Normative", "Complete") & cell_type == currentcell) %>% dplyr::select(gene, comparison)
  gene_list <- as.data.frame(table(gene_list$gene, gene_list$comparison))
  gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame(check.names = FALSE)
  gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)

  set_vars <- c("Gonad:addY", "Gonad:addX", "Gonad:SexChrom", "addY", "addX", "XX_vs_XY", "Ovary_vs_Testis", "XXF_vs_XYM")
  gene_list <- gene_list %>% dplyr::select(Var1, any_of(set_vars))
  
  plot <- ComplexUpset::upset(gene_list, names(gene_list)[-1], sort_sets = FALSE, sort_intersections = FALSE, name = currentcell,
                              themes=upset_default_themes(text=element_text(size=30)))
  
  plot_list[[currentcell]] <- plot
}

pdf("Plots/DEG/global/upset/Complete_overlap.pdf", width  = 70, height = 30)
do.call(ggarrange, c(plot_list, list(ncol = 5, nrow = 3)))
dev.off()
```

```{r}
##get barplot of just normative difference intersections - MAST output
DEG_df_filtered <- MAST_df %>% dplyr::filter(comb_p_val_adj < 0.05) %>% 
  mutate(model = factor(model, levels = c("Normative", "Gonad", "SexChrom", "addX", "addY1", "addY2")),
         cell_type = factor(cell_type))

plot_list <- list()
for(i in 1:nlevels(DEG_df_filtered$cell_type)) {
  
  currentcell <- levels(DEG_df_filtered$cell_type)[i]
  
  gene_list <- DEG_df_filtered %>% dplyr::filter(cell_type == currentcell) %>% dplyr::select(gene, model)
  gene_list <- as.data.frame(table(gene_list$gene, gene_list$model))
  gene_list <- gene_list %>% pivot_wider(names_from = Var2, values_from = Freq) %>% as.data.frame(check.names = FALSE)
  gene_list[-1] <- ifelse(gene_list[-1] == 0, 0, 1)

  set_vars <- c("addY2", "addY1", "addX", "SexChrom", "Gonad", "Normative")
  gene_list <- gene_list %>% dplyr::select(Var1, all_of(set_vars))
  plot <- ComplexUpset::upset(gene_list, names(gene_list)[-1], sort_sets = FALSE, sort_intersections = FALSE, name = currentcell,
                              themes=upset_default_themes(text=element_text(size=30)))
  
  plot_list[[currentcell]] <- plot
}

pdf("Plots/DEG/MAST/upset/model_overlap.pdf", width  = 50, height = 30)
do.call(ggarrange, c(plot_list, list(ncol = 5, nrow = 3)))
dev.off()
```

####################################################

